// @sanity/schema/lib/legacy/preview/guessPreviewConfig.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _isUndefined2 = require('lodash/isUndefined');

var _isUndefined3 = _interopRequireDefault(_isUndefined2);

var _omitBy2 = require('lodash/omitBy');

var _omitBy3 = _interopRequireDefault(_omitBy2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = guessPreviewFields;

var _arrify = require('arrify');

var _arrify2 = _interopRequireDefault(_arrify);

var _fallbackPrepare = require('./fallbackPrepare');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TITLE_CANDIDATES = ['title', 'name', 'label', 'heading', 'header', 'caption'];
var DESCRIPTION_CANDIDATES = ['description'].concat(TITLE_CANDIDATES);

function fieldHasReferenceTo(fieldDef, refType) {
  return (0, _arrify2.default)(fieldDef.to || []).some(function (memberTypeDef) {
    return memberTypeDef.type === refType;
  });
}

function isImageAssetField(fieldDef) {
  return fieldHasReferenceTo(fieldDef, 'sanity.imageAsset');
}

function resolveImageAssetPath(typeDef) {
  var fields = typeDef.fields || [];
  var imageAssetField = fields.find(isImageAssetField);
  if (imageAssetField) {
    return imageAssetField.name;
  }
  var fieldWithImageAsset = fields.find(function (fieldDef) {
    return (fieldDef.fields || []).some(isImageAssetField);
  });

  return fieldWithImageAsset ? fieldWithImageAsset.name + '.asset' : undefined;
}

function isFileAssetField(fieldDef) {
  return fieldHasReferenceTo(fieldDef, 'sanity.fileAsset');
}

function resolveFileAssetPath(typeDef) {
  var fields = typeDef.fields || [];
  var assetField = fields.find(isFileAssetField);
  if (assetField) {
    return assetField.name;
  }
  var fieldWithFileAsset = fields.find(function (fieldDef) {
    return (fieldDef.fields || []).some(isFileAssetField);
  });
  return fieldWithFileAsset ? fieldWithFileAsset.name + '.asset' : undefined;
}

function guessPreviewFields(rawObjectTypeDef) {
  var objectTypeDef = _extends({ fields: [] }, rawObjectTypeDef);

  var stringFieldNames = objectTypeDef.fields.filter(function (field) {
    return field.type === 'string';
  }).map(function (field) {
    return field.name;
  });

  // Check if we have fields with names that is listed in candidate fields
  var titleField = TITLE_CANDIDATES.find(function (candidate) {
    return stringFieldNames.includes(candidate);
  });

  var descField = DESCRIPTION_CANDIDATES.find(function (candidate) {
    return candidate !== titleField && stringFieldNames.includes(candidate);
  });

  if (!titleField) {
    // Pick first defined string field
    titleField = stringFieldNames[0];
    // Pick next as desc
    descField = stringFieldNames[1];
  }

  var imageAssetPath = resolveImageAssetPath(objectTypeDef);

  if (!titleField) {
    var fileAssetPath = resolveFileAssetPath(objectTypeDef);
    if (fileAssetPath) {
      titleField = fileAssetPath + '.originalFilename';
    }
    if (imageAssetPath) {
      titleField = imageAssetPath + '.originalFilename';
    }
  }

  if (!titleField && !imageAssetPath) {
    // last resort, pick all fields and concat them
    var fieldNames = objectTypeDef.fields.map(function (field) {
      return field.name;
    });
    var fieldMapping = fieldNames.reduce(function (acc, fieldName) {
      acc[fieldName] = fieldName;
      return acc;
    }, {});

    return {
      select: fieldMapping,
      prepare: (0, _fallbackPrepare.createFallbackPrepare)(fieldNames)
    };
  }

  var select = (0, _omitBy3.default)({
    title: titleField,
    description: descField,
    imageUrl: imageAssetPath ? imageAssetPath + '.url' : undefined
  }, _isUndefined3.default);

  return {
    select: select
  };
}
