// @sanity/core/lib/commands/hook/deleteHookCommand.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

let promptForHook = (() => {
  var _ref2 = _asyncToGenerator(function* (specified, context) {
    const specifiedName = specified && specified.toLowerCase();
    const prompt = context.prompt,
          apiClient = context.apiClient;

    const client = apiClient();

    const hooks = yield client.request({ uri: '/hooks', json: true });
    if (specifiedName) {
      const selected = hooks.filter(function (hook) {
        return hook.name.toLowerCase() === specifiedName;
      })[0];
      if (!selected) {
        throw new Error(`Hook with name "${specified} not found"`);
      }

      return selected.id;
    }

    const choices = hooks.map(function (hook) {
      return { value: hook.id, name: hook.name };
    });
    return prompt.single({
      message: 'Select hook to delete',
      type: 'list',
      choices: choices
    });
  });

  return function promptForHook(_x3, _x4) {
    return _ref2.apply(this, arguments);
  };
})();

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

exports.default = {
  name: 'delete',
  group: 'hook',
  signature: '[NAME]',
  description: 'Delete a hook within your project',
  action: (() => {
    var _ref = _asyncToGenerator(function* (args, context) {
      const apiClient = context.apiClient;

      var _args$argsWithoutOpti = _slicedToArray(args.argsWithoutOptions, 1);

      const name = _args$argsWithoutOpti[0];

      const client = apiClient();

      const hookId = yield promptForHook(name, context);
      try {
        yield client.request({ method: 'DELETE', uri: `/hooks/${hookId}` });
      } catch (err) {
        throw new Error(`Hook deletion failed:\n${err.message}`);
      }
    });

    return function action(_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })()
};
