// @sanity/components/lib/selects/CustomSelect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _customStyle = require('part:@sanity/components/selects/custom-style');

var _customStyle2 = _interopRequireDefault(_customStyle);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _angleDownIcon = require('part:@sanity/base/angle-down-icon');

var _angleDownIcon2 = _interopRequireDefault(_angleDownIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var CustomSelect = function (_React$Component) {
  _inherits(CustomSelect, _React$Component);

  function CustomSelect() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, CustomSelect);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = CustomSelect.__proto__ || Object.getPrototypeOf(CustomSelect)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      isOpen: false,
      activeIndex: -1
    }, _this.handleItemClick = function (event) {
      _this.selectIndex(Number(event.currentTarget.getAttribute('data-item-index')));
      _this.setState({ isOpen: false });
    }, _this.handleInnerClick = function () {
      _this.setState({ isOpen: true });
    }, _this.handleKeyDown = function (event) {
      var _this$state = _this.state,
          isOpen = _this$state.isOpen,
          activeIndex = _this$state.activeIndex;

      if (event.key === 'Enter') {
        if (isOpen && activeIndex > -1) {
          _this.selectIndex(activeIndex);
          _this.setState({ isOpen: false });
        } else {
          _this.setState({ isOpen: true });
        }
      }

      if (event.key === 'Escape') {
        _this.setState({ isOpen: false });
      }

      var items = _this.props.items;


      var lastIndex = items.length - 1;
      if (['ArrowUp', 'ArrowDown'].includes(event.key)) {
        event.preventDefault();
        if (!isOpen) {
          _this.setState({ isOpen: true });
          return;
        }
        var nextIndex = activeIndex + (event.key === 'ArrowUp' ? -1 : 1);
        if (nextIndex < 0) nextIndex = lastIndex;
        if (nextIndex > lastIndex) nextIndex = 0;
        _this.setState({ activeIndex: nextIndex });
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(CustomSelect, [{
    key: 'selectIndex',
    value: function selectIndex(index) {
      var item = this.props.items[index];
      this.props.onChange(item);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          items = _props.items,
          value = _props.value,
          renderItem = _props.renderItem,
          rest = _objectWithoutProperties(_props, ['items', 'value', 'renderItem']);

      var _state = this.state,
          isOpen = _state.isOpen,
          activeIndex = _state.activeIndex;


      return _react2.default.createElement(
        'div',
        _extends({ className: _customStyle2.default.root }, rest),
        _react2.default.createElement(
          'div',
          {
            className: _customStyle2.default.inner,
            onClick: this.handleInnerClick,
            onKeyDown: this.handleKeyDown,
            tabIndex: 0
          },
          _react2.default.createElement(
            'div',
            { className: _customStyle2.default.selectContainer },
            _react2.default.createElement(
              'span',
              { className: _customStyle2.default.text },
              renderItem(value)
            ),
            _react2.default.createElement(
              'div',
              { className: _customStyle2.default.arrow },
              _react2.default.createElement(_angleDownIcon2.default, { color: 'inherit' })
            )
          )
        ),
        isOpen && _react2.default.createElement(
          'div',
          { className: _customStyle2.default.listContainer },
          _react2.default.createElement(
            'ul',
            {
              className: _customStyle2.default.items,
              onKeyDown: this.handleKeyDown,
              ref: this.setListContainer
            },
            items.map(function (item, index) {
              var _cx;

              var isActive = index === activeIndex;
              var isSelected = item === value;
              var className = (0, _classnames2.default)(_customStyle2.default.item, (_cx = {}, _defineProperty(_cx, _customStyle2.default.selectedItem, isSelected), _defineProperty(_cx, _customStyle2.default.activeItem, isActive), _cx));
              return _react2.default.createElement(
                'li',
                {
                  key: index,
                  className: className,
                  onClick: _this2.handleItemClick, 'data-item-index': index
                },
                renderItem(item, { index: index, isActive: isActive, isSelected: isSelected })
              );
            })
          )
        )
      );
    }
  }]);

  return CustomSelect;
}(_react2.default.Component);

CustomSelect.propTypes = {
  onChange: _propTypes2.default.func,
  value: _propTypes2.default.object,
  renderItem: _propTypes2.default.func,
  items: _propTypes2.default.array
};
CustomSelect.defaultProps = {
  onChange: function onChange() {}
};
exports.default = CustomSelect;
