// @sanity/components/lib/selects/DefaultSelect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _defaultStyle = require('part:@sanity/components/selects/default-style');

var _defaultStyle2 = _interopRequireDefault(_defaultStyle);

var _angleDownIcon = require('part:@sanity/base/angle-down-icon');

var _angleDownIcon2 = _interopRequireDefault(_angleDownIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var DefaultSelect = function (_React$Component) {
  _inherits(DefaultSelect, _React$Component);

  function DefaultSelect() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, DefaultSelect);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = DefaultSelect.__proto__ || Object.getPrototypeOf(DefaultSelect)).call.apply(_ref, [this].concat(args))), _this), _this.handleChange = function (event) {
      _this.props.onChange(_this.props.items[event.target.value]);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(DefaultSelect, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          hasError = _props.hasError,
          items = _props.items,
          value = _props.value,
          disabled = _props.disabled,
          hasFocus = _props.hasFocus,
          rest = _objectWithoutProperties(_props, ['hasError', 'items', 'value', 'disabled', 'hasFocus']);

      return _react2.default.createElement(
        'div',
        {
          className: '\n        ' + (disabled ? _defaultStyle2.default.selectContainerDisabled : _defaultStyle2.default.selectContainer) + '\n        ' + (hasFocus ? _defaultStyle2.default.hasFocus : '') + '\n        ' + (hasError ? _defaultStyle2.default.hasError : '') + '\n      '
        },
        _react2.default.createElement(
          'select',
          _extends({}, rest, {
            className: _defaultStyle2.default.select,
            onChange: this.handleChange,
            disabled: disabled,
            value: value && items.indexOf(value),
            autoComplete: 'off'
          }),
          !value && _react2.default.createElement('option', null),
          items && items.length > 0 && items.map(function (item, i) {
            return _react2.default.createElement(
              'option',
              { key: i, value: i },
              item.title
            );
          })
        ),
        _react2.default.createElement('div', { className: _defaultStyle2.default.focusHelper }),
        _react2.default.createElement(
          'div',
          { className: _defaultStyle2.default.icon },
          _react2.default.createElement(_angleDownIcon2.default, { color: 'inherit' })
        )
      );
    }
  }]);

  return DefaultSelect;
}(_react2.default.Component);

DefaultSelect.propTypes = {
  onChange: _propTypes2.default.func,
  value: _propTypes2.default.object,
  hasError: _propTypes2.default.bool,
  onFocus: _propTypes2.default.func,
  onBlur: _propTypes2.default.func,
  hasFocus: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  items: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    title: _propTypes2.default.string
  }))
};
DefaultSelect.defaultProps = {
  onChange: function onChange() {},
  onBlur: function onBlur() {},
  onFocus: function onFocus() {},

  hasError: false,
  hasFocus: false,
  value: {},
  items: []
};
exports.default = DefaultSelect;
