// @sanity/components/lib/selects/RadioSelect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _radioStyle = require('part:@sanity/components/selects/radio-style');

var _radioStyle2 = _interopRequireDefault(_radioStyle);

var _default = require('part:@sanity/components/radiobutton/default');

var _default2 = _interopRequireDefault(_default);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RadioSelect = function (_React$Component) {
  _inherits(RadioSelect, _React$Component);

  function RadioSelect() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, RadioSelect);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = RadioSelect.__proto__ || Object.getPrototypeOf(RadioSelect)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      focusedItem: null
    }, _this.handleRadioChange = function (item) {
      _this.props.onChange(item);
    }, _this.handleFocus = function (item) {
      _this.setState({
        focusedItem: item
      });
    }, _this.handleBlur = function () {
      _this.setState({
        focusedItem: null
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(RadioSelect, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          items = _props.items,
          value = _props.value,
          name = _props.name,
          direction = _props.direction;
      var focusedItem = this.state.focusedItem;


      return _react2.default.createElement(
        'div',
        {
          className: '\n          ' + (direction == 'vertical' ? _radioStyle2.default.vertical : _radioStyle2.default.horizontal) + '\n        '
        },
        _react2.default.createElement(
          'div',
          { className: _radioStyle2.default.radioContainer },
          items.map(function (item, i) {
            return _react2.default.createElement(
              'div',
              { className: _radioStyle2.default.item, key: i },
              _react2.default.createElement(_default2.default, {
                name: name,
                key: i,
                label: item.title,
                item: item,
                onChange: _this2.handleRadioChange,
                checked: value === item,
                onFocus: _this2.handleFocus,
                onBlur: _this2.handleBlur,
                hasFocus: focusedItem === item
              })
            );
          })
        )
      );
    }
  }]);

  return RadioSelect;
}(_react2.default.Component);

RadioSelect.propTypes = {
  name: _propTypes2.default.string,
  direction: _propTypes2.default.oneOf(['horizontal', 'vertical']),
  onChange: _propTypes2.default.func,
  value: _propTypes2.default.object,
  items: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    title: _propTypes2.default.string
  }))
};
RadioSelect.defaultProps = {
  onChange: function onChange() {}
};
exports.default = RadioSelect;
