// @sanity/components/lib/selects/StatelessSearchableSelect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _searchableStyle = require('part:@sanity/components/selects/searchable-style');

var _searchableStyle2 = _interopRequireDefault(_searchableStyle);

var _angleDownIcon = require('part:@sanity/base/angle-down-icon');

var _angleDownIcon2 = _interopRequireDefault(_angleDownIcon);

var _default = require('part:@sanity/components/textinputs/default');

var _default2 = _interopRequireDefault(_default);

var _spinner = require('part:@sanity/components/loading/spinner');

var _spinner2 = _interopRequireDefault(_spinner);

var _closeIcon = require('part:@sanity/base/close-icon');

var _closeIcon2 = _interopRequireDefault(_closeIcon);

var _SelectMenu = require('./SelectMenu');

var _SelectMenu2 = _interopRequireDefault(_SelectMenu);

var _sticky = require('part:@sanity/components/portal/sticky');

var _sticky2 = _interopRequireDefault(_sticky);

var _Stacked = require('../utilities/Stacked');

var _Stacked2 = _interopRequireDefault(_Stacked);

var _CaptureOutsideClicks = require('../utilities/CaptureOutsideClicks');

var _CaptureOutsideClicks2 = _interopRequireDefault(_CaptureOutsideClicks);

var _Escapable = require('../utilities/Escapable');

var _Escapable2 = _interopRequireDefault(_Escapable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var noop = function noop() {};

var StatelessSearchableSelect = function (_React$PureComponent) {
  _inherits(StatelessSearchableSelect, _React$PureComponent);

  function StatelessSearchableSelect() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, StatelessSearchableSelect);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = StatelessSearchableSelect.__proto__ || Object.getPrototypeOf(StatelessSearchableSelect)).call.apply(_ref, [this].concat(args))), _this), _this.handleSelect = function (item) {
      _this.props.onChange(item);
    }, _this.handlePortalOutsideClick = function (event) {
      if (!_this._rootNode.contains(event.target)) {
        _this.props.onClose();
      }
    }, _this.handleArrowClick = function () {
      var _this$props = _this.props,
          isOpen = _this$props.isOpen,
          onOpen = _this$props.onOpen,
          onClose = _this$props.onClose;

      if (isOpen) {
        onClose();
      } else {
        onOpen();
      }
    }, _this.handleInputChange = function (event) {
      _this.props.onInputChange(event.target.value);
    }, _this.handleKeyDown = function (event) {
      var _this$props2 = _this.props,
          items = _this$props2.items,
          highlightIndex = _this$props2.highlightIndex,
          onHighlightIndexChange = _this$props2.onHighlightIndexChange,
          isOpen = _this$props2.isOpen,
          onOpen = _this$props2.onOpen;

      if (!items || items.length === 0) {
        return;
      }

      var lastIndex = items.length - 1;
      if (event.key === 'ArrowUp') {
        event.preventDefault();
        var nextIndex = highlightIndex - 1;
        onHighlightIndexChange(nextIndex < 0 ? lastIndex : nextIndex);
      }
      if (event.key === 'ArrowDown') {
        event.preventDefault();
        if (!isOpen) {
          onOpen();
        }
        var _nextIndex = highlightIndex + 1;
        onHighlightIndexChange(_nextIndex > lastIndex ? 0 : _nextIndex);
      }
    }, _this.handleKeyUp = function (event) {
      var _this$props3 = _this.props,
          items = _this$props3.items,
          onChange = _this$props3.onChange,
          highlightIndex = _this$props3.highlightIndex;

      if (event.key === 'Enter' && highlightIndex > -1 && items[highlightIndex]) {
        onChange(items[highlightIndex]);
      }
    }, _this.setListElement = function (element) {
      _this._listElement = element;
    }, _this.setRootNode = function (node) {
      _this._rootNode = node;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(StatelessSearchableSelect, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          onClear = _props.onClear,
          placeholder = _props.placeholder,
          isLoading = _props.isLoading,
          value = _props.value,
          items = _props.items,
          renderItem = _props.renderItem,
          isOpen = _props.isOpen,
          highlightIndex = _props.highlightIndex,
          isInputSelected = _props.isInputSelected,
          inputValue = _props.inputValue,
          onChange = _props.onChange,
          onInputChange = _props.onInputChange,
          onOpen = _props.onOpen,
          onClose = _props.onClose,
          onResize = _props.onResize,
          dropdownPosition = _props.dropdownPosition,
          disabled = _props.disabled,
          onHighlightIndexChange = _props.onHighlightIndexChange,
          openItemElement = _props.openItemElement,
          rest = _objectWithoutProperties(_props, ['onClear', 'placeholder', 'isLoading', 'value', 'items', 'renderItem', 'isOpen', 'highlightIndex', 'isInputSelected', 'inputValue', 'onChange', 'onInputChange', 'onOpen', 'onClose', 'onResize', 'dropdownPosition', 'disabled', 'onHighlightIndexChange', 'openItemElement']);

      return _react2.default.createElement(
        'div',
        { ref: this.setRootNode },
        _react2.default.createElement(
          'div',
          { className: disabled ? _searchableStyle2.default.selectContainerDisabled : _searchableStyle2.default.selectContainer },
          _react2.default.createElement(_default2.default, _extends({}, rest, {
            className: _searchableStyle2.default.select,
            placeholder: placeholder,
            onChange: this.handleInputChange,
            onKeyDown: this.handleKeyDown,
            onKeyUp: this.handleKeyUp,
            value: inputValue || '',
            selected: isInputSelected,
            disabled: disabled
          })),
          onClear && inputValue && _react2.default.createElement(
            'button',
            { className: _searchableStyle2.default.clearButton, onClick: onClear },
            _react2.default.createElement(_closeIcon2.default, { color: 'inherit' })
          ),
          isLoading && _react2.default.createElement(
            'div',
            { className: _searchableStyle2.default.spinner },
            _react2.default.createElement(_spinner2.default, null)
          ),
          openItemElement && value && _react2.default.createElement(
            'span',
            { className: _searchableStyle2.default.openItem },
            openItemElement(value)
          ),
          !isLoading && _react2.default.createElement(
            'div',
            { className: _searchableStyle2.default.arrow, onClick: disabled ? null : this.handleArrowClick },
            _react2.default.createElement(_angleDownIcon2.default, { color: 'inherit' })
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _searchableStyle2.default.stickyContainer, style: { width: this.props.width + 'px' } },
          isOpen && _react2.default.createElement(
            _sticky2.default,
            {
              isOpen: isOpen,
              onResize: onResize,
              onlyBottomSpace: false,
              useOverlay: false,
              addPadding: false,
              scrollIntoView: false
            },
            _react2.default.createElement(
              _Stacked2.default,
              null,
              function (isActive) {
                return _react2.default.createElement(
                  'div',
                  {
                    className: '\n                        ' + (isOpen ? _searchableStyle2.default.listContainer : _searchableStyle2.default.listContainerHidden) + '\n                        ' + (dropdownPosition === 'top' ? _searchableStyle2.default.listContainerTop : _searchableStyle2.default.listContainerBottom) + '\n                        ' + (items.length === 0 ? _searchableStyle2.default.listContainerEmpty : '') + '\n                      ',
                    style: { width: _this2.props.width + 'px' },
                    ref: _this2.setListElement
                  },
                  items.length === 0 && !isLoading && _react2.default.createElement(
                    'p',
                    { className: _searchableStyle2.default.noResultText },
                    'No results'
                  ),
                  items.length === 0 && isLoading && _react2.default.createElement(
                    'div',
                    { className: _searchableStyle2.default.listSpinner },
                    _react2.default.createElement(_spinner2.default, { message: 'Loading items\u2026' })
                  ),
                  items.length > 0 && _react2.default.createElement(
                    _CaptureOutsideClicks2.default,
                    { onClickOutside: isActive ? onClose : null },
                    _react2.default.createElement(_Escapable2.default, { onEscape: function onEscape(event) {
                        return (isActive || event.shiftKey) && onClose();
                      } }),
                    _react2.default.createElement(_SelectMenu2.default, {
                      items: items,
                      value: value,
                      onSelect: _this2.handleSelect,
                      renderItem: renderItem,
                      highlightIndex: highlightIndex
                    })
                  )
                );
              }
            )
          )
        )
      );
    }
  }]);

  return StatelessSearchableSelect;
}(_react2.default.PureComponent);

StatelessSearchableSelect.propTypes = {
  onChange: _propTypes2.default.func,
  value: _propTypes2.default.any,
  inputValue: _propTypes2.default.string,
  onInputChange: _propTypes2.default.func,
  onClear: _propTypes2.default.func,
  renderItem: _propTypes2.default.func,
  placeholder: _propTypes2.default.string,
  isLoading: _propTypes2.default.bool,
  isOpen: _propTypes2.default.bool,
  onOpen: _propTypes2.default.func,
  onClose: _propTypes2.default.func,
  onResize: _propTypes2.default.func,
  openItemElement: _propTypes2.default.func,
  items: _propTypes2.default.array,
  highlightIndex: _propTypes2.default.number,
  onHighlightIndexChange: _propTypes2.default.func,
  isInputSelected: _propTypes2.default.bool,
  width: _propTypes2.default.number,
  disabled: _propTypes2.default.bool,
  dropdownPosition: _propTypes2.default.string
};
StatelessSearchableSelect.defaultProps = {
  onChange: noop,
  onOpen: noop,
  onClose: noop,
  onInputChange: noop,
  isLoading: false,
  renderItem: function renderItem(item) {
    return item;
  },
  items: [],
  width: 100,
  dropdownPosition: 'bottom'
};
exports.default = StatelessSearchableSelect;
