// @sanity/components/lib/buttons/DefaultButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactInk = require('react-ink');

var _reactInk2 = _interopRequireDefault(_reactInk);

var _defaultStyle = require('part:@sanity/components/buttons/default-style');

var _defaultStyle2 = _interopRequireDefault(_defaultStyle);

var _spinner = require('part:@sanity/components/loading/spinner');

var _spinner2 = _interopRequireDefault(_spinner);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var DefaultButton = function (_React$Component) {
  _inherits(DefaultButton, _React$Component);

  function DefaultButton() {
    var _ref;

    _classCallCheck(this, DefaultButton);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var _this = _possibleConstructorReturn(this, (_ref = DefaultButton.__proto__ || Object.getPrototypeOf(DefaultButton)).call.apply(_ref, [this].concat(args)));

    _this.handleClick = _this.handleClick.bind(_this);

    _this.state = {
      recentlyHovered: false
    };
    return _this;
  }

  _createClass(DefaultButton, [{
    key: 'handleClick',
    value: function handleClick(event) {
      var _this2 = this;

      this.setState({
        recentlyHovered: true
      });
      this.props.onClick(event);
      clearTimeout(this.timeOut);
      this.timeOut = setTimeout(function () {
        _this2.setState({
          recentlyHovered: true
        });
      }, 300);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      clearTimeout(this.timeOut);
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          kind = _props.kind,
          ripple = _props.ripple,
          disabled = _props.disabled,
          inverted = _props.inverted,
          color = _props.color,
          icon = _props.icon,
          loading = _props.loading,
          className = _props.className,
          children = _props.children,
          rest = _objectWithoutProperties(_props, ['kind', 'ripple', 'disabled', 'inverted', 'color', 'icon', 'loading', 'className', 'children']);

      var Icon = icon;

      var style = [inverted && _defaultStyle2.default.inverted, kind ? _defaultStyle2.default[kind] : _defaultStyle2.default.default, color && _defaultStyle2.default['color__' + color], icon && _defaultStyle2.default.hasIcon, className, disabled && _defaultStyle2.default.disabled, this.state.recentlyHovered ? _defaultStyle2.default.recentlyHovered : _defaultStyle2.default.notRecentlyHovered].filter(Boolean).join(' ');

      var padContent = Icon && children && typeof children === 'string';

      return _react2.default.createElement(
        'button',
        _extends({}, rest, {
          className: style,
          onClick: this.handleClick,
          disabled: disabled
        }),
        _react2.default.createElement(
          'div',
          { className: _defaultStyle2.default.inner },
          loading && _react2.default.createElement(
            'span',
            { className: _defaultStyle2.default.spinner },
            _react2.default.createElement(_spinner2.default, { inline: true })
          ),
          Icon && _react2.default.createElement(Icon, { className: _defaultStyle2.default.icon }),
          children && _react2.default.createElement(
            'span',
            { className: padContent ? _defaultStyle2.default.contentWithPad : _defaultStyle2.default.content },
            children
          ),
          ripple && !disabled && _react2.default.createElement(_reactInk2.default, { duration: 200, opacity: 0.10, radius: 200 })
        )
      );
    }
  }]);

  return DefaultButton;
}(_react2.default.Component);

DefaultButton.propTypes = {
  kind: _propTypes2.default.oneOf(['default', 'simple', 'secondary']),
  color: _propTypes2.default.oneOf(['primary', 'success', 'danger', 'white']),
  onClick: _propTypes2.default.func,
  children: _propTypes2.default.node,
  inverted: _propTypes2.default.bool,
  icon: _propTypes2.default.func,
  loading: _propTypes2.default.bool,
  ripple: _propTypes2.default.bool,
  className: _propTypes2.default.string,
  disabled: _propTypes2.default.bool,
  tabIndex: _propTypes2.default.number,
  type: _propTypes2.default.string
};
DefaultButton.defaultProps = {
  ripple: true,
  icon: null,
  onClick: function onClick() {},

  kind: 'default',
  type: 'button'
};
exports.default = DefaultButton;
