// @sanity/components/lib/toggles/Checkbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _checkboxStyle = require('part:@sanity/components/toggles/checkbox-style');

var _checkboxStyle2 = _interopRequireDefault(_checkboxStyle);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Checkbox = function (_React$Component) {
  _inherits(Checkbox, _React$Component);

  function Checkbox() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Checkbox);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Checkbox.__proto__ || Object.getPrototypeOf(Checkbox)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      hasFocus: false
    }, _this.handleMouseUp = function () {
      _this.setState({ hasFocus: false });
    }, _this.handleFocus = function () {
      _this.setState({ hasFocus: true });
    }, _this.handleChange = function (event) {
      _this.props.onChange(event, _this.props.value);
    }, _this.handleBlur = function () {
      window.setTimeout(function () {
        _this.props.onBlur(_this.props.value);
      }, 0.001);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Checkbox, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          disabled = _props.disabled,
          hasFocus = _props.hasFocus,
          value = _props.value,
          checked = _props.checked,
          label = _props.label,
          children = _props.children;


      return _react2.default.createElement(
        'label',
        {
          title: label,
          className: '\n          ' + _checkboxStyle2.default.root + '\n          ' + (disabled ? _checkboxStyle2.default.isDisabled : _checkboxStyle2.default.isEnabled) + '\n          ' + (checked ? _checkboxStyle2.default.isChecked : _checkboxStyle2.default.unChecked) + '\n          ' + (hasFocus ? _checkboxStyle2.default.hasFocus : '') + '\n        ',
          onMouseUp: this.handleMouseUp
        },
        _react2.default.createElement('input', {
          className: _checkboxStyle2.default.input,
          type: 'checkbox',
          value: String(value),
          onChange: this.handleChange,
          checked: checked,
          onFocus: this.handleFocus,
          onBlur: this.handleBlur
        }),
        _react2.default.createElement(
          'div',
          { className: _checkboxStyle2.default.label },
          children || label
        ),
        _react2.default.createElement('div', { className: _checkboxStyle2.default.focusHelper }),
        _react2.default.createElement(
          'div',
          { className: _checkboxStyle2.default.boxOutline },
          _react2.default.createElement('div', { className: _checkboxStyle2.default.tickOutline })
        ),
        _react2.default.createElement('div', { className: _checkboxStyle2.default.tickHelper }),
        _react2.default.createElement('div', { className: _checkboxStyle2.default.rippleContainer })
      );
    }
  }]);

  return Checkbox;
}(_react2.default.Component);

Checkbox.propTypes = {
  value: _propTypes2.default.bool,
  label: _propTypes2.default.string,
  onChange: _propTypes2.default.func,
  onFocus: _propTypes2.default.func,
  onBlur: _propTypes2.default.func,
  checked: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  hasFocus: _propTypes2.default.bool,
  children: _propTypes2.default.node
};
Checkbox.defaultProps = {
  onChange: function onChange() {},
  onFocus: function onFocus() {},
  onBlur: function onBlur() {}
};
exports.default = Checkbox;
