// @sanity/components/lib/toggles/Switch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _switchStyle = require('part:@sanity/components/toggles/switch-style');

var _switchStyle2 = _interopRequireDefault(_switchStyle);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Switch = function (_React$Component) {
  _inherits(Switch, _React$Component);

  function Switch() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Switch);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Switch.__proto__ || Object.getPrototypeOf(Switch)).call.apply(_ref, [this].concat(args))), _this), _this.handleMouseUp = function (event) {
      _this.handleBlur();
    }, _this.handleBlur = function (event) {
      window.setTimeout(function () {
        _this.props.onBlur();
      }, 0.001);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Switch, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          disabled = _props.disabled,
          checked = _props.checked,
          onChange = _props.onChange,
          onFocus = _props.onFocus,
          hasFocus = _props.hasFocus;


      return _react2.default.createElement(
        'label',
        {
          className: '\n          ' + (disabled ? _switchStyle2.default.disabled : _switchStyle2.default.enabled) + '\n          ' + (checked ? _switchStyle2.default.isChecked : _switchStyle2.default.unchecked) + '\n          ' + (hasFocus ? _switchStyle2.default.hasFocus : '') + '\n        ',
          onMouseUp: this.handleMouseUp
        },
        _react2.default.createElement('div', { className: _switchStyle2.default.track }),
        _react2.default.createElement(
          'div',
          { className: '' + (checked ? _switchStyle2.default.thumbChecked : _switchStyle2.default.thumb) },
          _react2.default.createElement('div', { className: _switchStyle2.default.focusHelper })
        ),
        _react2.default.createElement('input', {
          className: _switchStyle2.default.input,
          type: 'checkbox',
          checked: checked,
          readOnly: disabled,
          onFocus: onFocus,
          onBlur: this.handleBlur,
          onChange: onChange
        }),
        _react2.default.createElement(
          'div',
          { className: _switchStyle2.default.label },
          this.props.label
        )
      );
    }
  }]);

  return Switch;
}(_react2.default.Component);

Switch.propTypes = {
  label: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func,
  onFocus: _propTypes2.default.func,
  onBlur: _propTypes2.default.func,
  checked: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  hasFocus: _propTypes2.default.bool
};
Switch.defaultProps = {
  onChange: function onChange() {},
  onBlur: function onBlur() {},
  onFocus: function onFocus() {}
};
exports.default = Switch;
