// @sanity/default-layout/lib/components/ActionModal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _router = require('part:@sanity/base/router');

var _fullscreen = require('part:@sanity/components/dialogs/fullscreen');

var _fullscreen2 = _interopRequireDefault(_fullscreen);

var _ActionModal = require('./styles/ActionModal.css');

var _ActionModal2 = _interopRequireDefault(_ActionModal);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ActionModal(props) {
  return _react2.default.createElement(
    _fullscreen2.default,
    { className: _ActionModal2.default.modal, title: props.title, onClose: props.onClose, isOpen: true },
    _react2.default.createElement(
      'div',
      { className: _ActionModal2.default.content },
      _react2.default.createElement(
        'ul',
        { className: _ActionModal2.default.list },
        props.actions.map(function (action) {
          return _react2.default.createElement(
            'li',
            { className: _ActionModal2.default.listItem, key: action.title },
            _react2.default.createElement(
              _router.IntentLink,
              {
                onClick: props.onClose,
                className: _ActionModal2.default.actionLink,
                intent: 'create',
                params: action.params
              },
              action.title
            )
          );
        })
      )
    )
  );
}

ActionModal.defaultProps = {
  title: 'Create',
  actions: []
};

ActionModal.propTypes = {
  title: _propTypes2.default.string,
  actions: _propTypes2.default.array,
  onClose: _propTypes2.default.func.isRequired
};

exports.default = ActionModal;
