// @sanity/default-layout/lib/components/DefaultLayoutContainer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _DefaultLayout = require('./DefaultLayout');

var _DefaultLayout2 = _interopRequireDefault(_DefaultLayout);

var _loginWrapper = require('part:@sanity/base/login-wrapper?');

var _loginWrapper2 = _interopRequireDefault(_loginWrapper);

var _router = require('part:@sanity/base/router');

var _NotFound = require('./NotFound');

var _NotFound2 = _interopRequireDefault(_NotFound);

var _getOrderedTools = require('../util/getOrderedTools');

var _getOrderedTools2 = _interopRequireDefault(_getOrderedTools);

var _defaultLayoutRouter = require('../defaultLayoutRouter');

var _defaultLayoutRouter2 = _interopRequireDefault(_defaultLayoutRouter);

var _urlState = require('../datastores/urlState');

var urlStateStore = _interopRequireWildcard(_urlState);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var DefaultLayoutContainer = function (_React$PureComponent) {
  _inherits(DefaultLayoutContainer, _React$PureComponent);

  function DefaultLayoutContainer() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, DefaultLayoutContainer);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = DefaultLayoutContainer.__proto__ || Object.getPrototypeOf(DefaultLayoutContainer)).call.apply(_ref, [this].concat(args))), _this), _this.state = {}, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(DefaultLayoutContainer, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this2 = this;

      this.urlStateSubscription = urlStateStore.state.subscribe({
        next: function next(event) {
          return _this2.setState({
            urlState: event.state,
            isNotFound: event.isNotFound,
            intent: event.intent
          });
        }
      });
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.urlStateSubscription.unsubscribe();
    }
  }, {
    key: 'handleNavigate',
    value: function handleNavigate(newUrl, options) {
      urlStateStore.navigate(newUrl, options);
    }
  }, {
    key: 'render',
    value: function render() {
      var _state = this.state,
          intent = _state.intent,
          urlState = _state.urlState,
          isNotFound = _state.isNotFound;

      var tools = (0, _getOrderedTools2.default)();

      var content = isNotFound ? _react2.default.createElement(
        _NotFound2.default,
        null,
        intent && _react2.default.createElement(
          'div',
          null,
          'No tool can handle the intent:',
          ' ',
          ' ',
          _react2.default.createElement(
            'strong',
            null,
            intent.name
          ),
          ' ',
          ' ',
          'with parameters ',
          _react2.default.createElement(
            'pre',
            null,
            JSON.stringify(intent.params)
          )
        )
      ) : _react2.default.createElement(_DefaultLayout2.default, { tools: tools });

      var router = _react2.default.createElement(
        _router.RouterProvider,
        { router: _defaultLayoutRouter2.default, state: urlState, onNavigate: this.handleNavigate },
        content
      );

      return _loginWrapper2.default ? _react2.default.createElement(
        _loginWrapper2.default,
        null,
        router
      ) : router;
    }
  }]);

  return DefaultLayoutContainer;
}(_react2.default.PureComponent);

exports.default = DefaultLayoutContainer;
