// @sanity/default-layout/lib/components/LoginStatus.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactClickOutside = require('react-click-outside');

var _reactClickOutside2 = _interopRequireDefault(_reactClickOutside);

var _user = require('part:@sanity/base/user');

var _user2 = _interopRequireDefault(_user);

var _default = require('part:@sanity/components/menus/default');

var _default2 = _interopRequireDefault(_default);

var _signOutIcon = require('part:@sanity/base/sign-out-icon');

var _signOutIcon2 = _interopRequireDefault(_signOutIcon);

var _LoginStatus = require('./styles/LoginStatus.css');

var _LoginStatus2 = _interopRequireDefault(_LoginStatus);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var LoginStatus = function (_React$PureComponent) {
  _inherits(LoginStatus, _React$PureComponent);

  function LoginStatus() {
    var _ref;

    _classCallCheck(this, LoginStatus);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var _this = _possibleConstructorReturn(this, (_ref = LoginStatus.__proto__ || Object.getPrototypeOf(LoginStatus)).call.apply(_ref, [this].concat(args)));

    _this.handleClickOutside = function () {
      if (_this.state.userMenuOpened) {
        _this.setState({ userMenuOpened: false });
      }
    };

    _this.handleUserMenuToggle = function () {
      _this.setState({ userMenuOpened: !_this.state.userMenuOpened });
    };

    _this.handleUserMenuItemClick = function (item) {
      if (item.index === 'signOut') {
        _user2.default.actions.logout();
      }
    };

    _this.state = {
      userMenuOpened: false
    };
    return _this;
  }

  _createClass(LoginStatus, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      var _this2 = this;

      this.userSubscription = _user2.default.currentUser.subscribe(function (event) {
        return _this2.setState({ user: event.user });
      });
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.userSubscription.unsubscribe();
    }
  }, {
    key: 'render',
    value: function render() {
      var className = this.props.className;
      var _state = this.state,
          user = _state.user,
          userMenuOpened = _state.userMenuOpened;

      if (!user) {
        return null;
      }

      return _react2.default.createElement(
        'div',
        { className: _LoginStatus2.default.root + ' ' + className },
        _react2.default.createElement(
          'div',
          { onClick: this.handleUserMenuToggle },
          _react2.default.createElement('img', { src: user.profileImage, className: _LoginStatus2.default.userImage })
        ),
        _react2.default.createElement(
          'div',
          { className: _LoginStatus2.default.userName },
          user.name
        ),
        userMenuOpened && _react2.default.createElement(
          'div',
          { className: _LoginStatus2.default.userMenu },
          _react2.default.createElement(_default2.default, {
            onAction: this.handleUserMenuItemClick,
            items: [{
              title: 'Log out ' + user.name,
              icon: _signOutIcon2.default,
              index: 'signOut'
            }],
            opened: true,
            origin: 'top-right',
            onClickOutside: this.handleUserMenuClose
          })
        )
      );
    }
  }]);

  return LoginStatus;
}(_react2.default.PureComponent);

LoginStatus.propTypes = {
  className: _propTypes2.default.string
};
exports.default = (0, _reactClickOutside2.default)(LoginStatus);
