// @sanity/form-builder/lib/utils/ValueSync.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _debounce2 = require('lodash/debounce');

var _debounce3 = _interopRequireDefault(_debounce2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _why = require('is-equal/why');

var _why2 = _interopRequireDefault(_why);

var _withPatchSubscriber = require('./withPatchSubscriber');

var _withPatchSubscriber2 = _interopRequireDefault(_withPatchSubscriber);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*:: import type {Patch} from '../utils/patches'*/
/*:: declare var __DEV__: boolean*/
/*:: type Deserialized = any*/
/*:: type ChildProps = { value: Deserialized }*/
/*:: type Props = {
  value: Object,
  subscribe: Function,
  serialize: (value: Deserialized) => Object,
  deserialize: (value: Object) => Deserialized,
  applyPatch: (patch: Patch) => Deserialized,
  children: (ChildProps) => ?React$Element<any>
}*/
exports.default = (0, _withPatchSubscriber2.default)((_temp = _class = function (_React$Component) {
  _inherits(ValueSync, _React$Component);

  function ValueSync(props /*: Props*/) {
    _classCallCheck(this, ValueSync);

    var _this = _possibleConstructorReturn(this, (ValueSync.__proto__ || Object.getPrototypeOf(ValueSync)).call(this));

    _this.checkDiff = (0, _debounce3.default)(function () {
      var propsVal = _this.props.value;
      var stateVal = _this.state.value ? _this.props.serialize(_this.state.value) : _this.state.value;
      var notEqual = (0, _why2.default)(propsVal, stateVal);
      if (notEqual) {
        // eslint-disable-next-line no-console
        console.warn('Serialized local input value (%o) out of sync with actual value (%o): %s', propsVal, stateVal, notEqual);
      }
    }, 5000);

    _this.state = {
      value: props.deserialize(props.value)
    };
    _this.unsubscribe = props.subscribe(function (_ref) {
      var snapshot = _ref.snapshot,
          patches = _ref.patches,
          shouldReset = _ref.shouldReset;

      if (shouldReset) {
        // eslint-disable-next-line no-console
        console.warn('Serialized local input value was reset due to a patch that targeted an ancestor');
        _this.setState({ value: props.deserialize(snapshot) });
      }
      _this.receivePatches(patches);
    });
    return _this;
  }

  _createClass(ValueSync, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.unsubscribe();
      this.checkDiff.cancel();
    }
  }, {
    key: 'receivePatches',
    value: function receivePatches(patches /*: Array<Patch>*/) {
      var applyPatch = this.props.applyPatch;

      this.setState(function (prevState) {
        return {
          value: patches.reduce(applyPatch, prevState.value)
        };
      });
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      if (__DEV__) {
        this.checkDiff();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var value = this.state.value;

      return this.props.children(_extends({}, this.props, { value: value }));
    }
  }]);

  return ValueSync;
}(_react2.default.Component), _class.contextTypes = {
  getValuePath: _propTypes2.default.func,
  formBuilder: _propTypes2.default.any
}, _temp));
