// @sanity/base/lib/datastores/user/createUserStore.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createUserStore;

var _observable = require('@sanity/observable');

var _observable2 = _interopRequireDefault(_observable);

var _createActions = require('../utils/createActions');

var _createActions2 = _interopRequireDefault(_createActions);

var _nanoPubsub = require('nano-pubsub');

var _nanoPubsub2 = _interopRequireDefault(_nanoPubsub);

var _authenticationFetcher = require('part:@sanity/base/authentication-fetcher');

var _authenticationFetcher2 = _interopRequireDefault(_authenticationFetcher);

var _client = require('part:@sanity/base/client');

var _client2 = _interopRequireDefault(_client);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var userChannel = (0, _nanoPubsub2.default)();
var errorChannel = (0, _nanoPubsub2.default)();

var _initialFetched = false;
var _currentUser = null;
var _currentError = null;

userChannel.subscribe(function (val) {
  _currentUser = val;
});

errorChannel.subscribe(function (val) {
  _currentError = val;
});

function fetchInitial() {
  return _authenticationFetcher2.default.getCurrentUser().then(function (user) {
    return userChannel.publish(user);
  }, function (err) {
    return errorChannel.publish(err);
  });
}

function logout() {
  return _authenticationFetcher2.default.logout().then(function () {
    return userChannel.publish(null);
  }, function (err) {
    return errorChannel.publish(err);
  });
}

var currentUser = new _observable2.default(function (observer) {
  if (_initialFetched) {
    var emitter = _currentError ? emitError : emitSnapshot;
    emitter(_currentError || _currentUser);
  } else {
    _initialFetched = true;
    fetchInitial();
  }

  var unsubUser = userChannel.subscribe(function (nextUser) {
    return emitSnapshot(nextUser);
  });
  var unsubError = errorChannel.subscribe(function (err) {
    return emitError(err);
  });
  var unsubscribe = function unsubscribe() {
    unsubUser();
    unsubError();
  };

  return unsubscribe;

  function emitError(error) {
    observer.next({ type: 'error', error: error });
  }

  function emitSnapshot(user) {
    observer.next({ type: 'snapshot', user: user });
  }
});

var userCache = {};

var getUser = function getUser(id) {

  if (!userCache[id]) {
    userCache[id] = _client2.default.request({
      uri: '/users/' + id,
      withCredentials: true
    }).then(function (user) {
      return user && user.id ? user : null;
    });
  }

  return userCache[id];
};

function createUserStore() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  return {
    actions: (0, _createActions2.default)({ logout: logout, retry: fetchInitial }),
    currentUser: currentUser,
    getUser: getUser
  };
}
