// @sanity/desk-tool/lib/components/ConfirmDelete.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _fullscreen = require('part:@sanity/components/dialogs/fullscreen');

var _fullscreen2 = _interopRequireDefault(_fullscreen);

var _spinner = require('part:@sanity/components/loading/spinner');

var _spinner2 = _interopRequireDefault(_spinner);

var _enhanceWithReferringDocuments = require('./enhanceWithReferringDocuments');

var _enhanceWithReferringDocuments2 = _interopRequireDefault(_enhanceWithReferringDocuments);

var _DocTitle = require('./DocTitle');

var _DocTitle2 = _interopRequireDefault(_DocTitle);

var _ReferringDocumentsList = require('./ReferringDocumentsList');

var _ReferringDocumentsList2 = _interopRequireDefault(_ReferringDocumentsList);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

exports.default = (0, _enhanceWithReferringDocuments2.default)((_temp2 = _class = function (_React$PureComponent) {
  _inherits(ConfirmDelete, _React$PureComponent);

  function ConfirmDelete() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ConfirmDelete);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ConfirmDelete.__proto__ || Object.getPrototypeOf(ConfirmDelete)).call.apply(_ref, [this].concat(args))), _this), _this.handleAction = function (action) {
      var _this$props = _this.props,
          onCancel = _this$props.onCancel,
          onConfirm = _this$props.onConfirm;

      if (action.name === 'confirm') {
        onConfirm();
      }
      if (action.name === 'cancel') {
        onCancel();
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ConfirmDelete, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          isCheckingReferringDocuments = _props.isCheckingReferringDocuments,
          referringDocuments = _props.referringDocuments,
          draft = _props.draft,
          published = _props.published,
          onCancel = _props.onCancel;


      var hasReferringDocuments = referringDocuments.length > 0;

      var canContinue = !isCheckingReferringDocuments;

      var actions = [canContinue && { name: 'confirm', title: hasReferringDocuments ? 'Try to delete anyway' : 'Delete now' }, { name: 'cancel', title: 'Cancel', kind: 'secondary' }].filter(Boolean);

      var title = isCheckingReferringDocuments ? 'Checking…' : 'Confirm delete';

      return _react2.default.createElement(
        _fullscreen2.default,
        {
          isOpen: true,
          showHeader: true,
          color: 'danger',
          centered: true,
          title: title,
          onClose: onCancel,
          onAction: this.handleAction,
          actions: actions
        },
        isCheckingReferringDocuments && _react2.default.createElement(_spinner2.default, { message: 'Looking for referring documents\u2026' }),
        hasReferringDocuments && _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(
            'h3',
            null,
            'Warning: Found ',
            referringDocuments.length === 1 ? 'a document' : referringDocuments.length + ' documents',
            ' that refers to ',
            '"',
            _react2.default.createElement(_DocTitle2.default, { document: draft || published }),
            '"'
          ),
          _react2.default.createElement(
            'p',
            null,
            'You may not be allowed to delete',
            ' "',
            _react2.default.createElement(_DocTitle2.default, { document: draft || published }),
            '" ',
            'as the following document',
            referringDocuments.length === 1 ? '' : 's',
            ' refers to it:'
          ),
          _react2.default.createElement(_ReferringDocumentsList2.default, { documents: referringDocuments })
        ),
        !isCheckingReferringDocuments && !hasReferringDocuments && _react2.default.createElement(
          'div',
          { style: { padding: 10 } },
          _react2.default.createElement(
            'p',
            null,
            'Are you sure you would like to delete the document',
            ' ',
            _react2.default.createElement(
              'strong',
              null,
              _react2.default.createElement(_DocTitle2.default, { document: draft || published })
            ),
            '?'
          ),
          _react2.default.createElement(
            'h2',
            null,
            'Careful!'
          ),
          _react2.default.createElement(
            'p',
            null,
            'If you continue, this document will be permanently gone and not be available to anyone anymore.'
          )
        )
      );
    }
  }]);

  return ConfirmDelete;
}(_react2.default.PureComponent), _class.propTypes = {
  onCancel: _propTypes2.default.func.isRequired,
  onConfirm: _propTypes2.default.func.isRequired,
  published: _propTypes2.default.object,
  draft: _propTypes2.default.object,
  referringDocuments: _propTypes2.default.array,
  isCheckingReferringDocuments: _propTypes2.default.bool
}, _temp2));
