// @sanity/components/lib/dialogs/DefaultDialog.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _closeIcon = require('part:@sanity/base/close-icon');

var _closeIcon2 = _interopRequireDefault(_closeIcon);

var _defaultStyle = require('part:@sanity/components/dialogs/default-style');

var _defaultStyle2 = _interopRequireDefault(_defaultStyle);

var _default = require('part:@sanity/components/buttons/default');

var _default2 = _interopRequireDefault(_default);

var _reactPortal = require('react-portal');

var _reactPortal2 = _interopRequireDefault(_reactPortal);

var _StackedEscapable = require('../utilities/StackedEscapable');

var _StackedEscapable2 = _interopRequireDefault(_StackedEscapable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var DefaultDialog = function (_React$PureComponent) {
  _inherits(DefaultDialog, _React$PureComponent);

  function DefaultDialog() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, DefaultDialog);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = DefaultDialog.__proto__ || Object.getPrototypeOf(DefaultDialog)).call.apply(_ref, [this].concat(args))), _this), _this.handleDialogClick = function (event) {
      event.stopPropagation();
    }, _this.handleActionClick = function (event) {
      var actionIndex = event.currentTarget.getAttribute('data-action-index');
      _this.props.onAction(_this.props.actions[actionIndex]);
    }, _this.setDialogElement = function (element) {
      _this.dialog = element;
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(DefaultDialog, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps) {
      var isOpen = this.props.isOpen;
      var wasOpen = prevProps.isOpen;
      if (!wasOpen && isOpen) {
        this.openDialogElement();
      } else if (wasOpen && !isOpen) {
        this.closeDialogElement();
      }
    }
  }, {
    key: 'openDialogElement',
    value: function openDialogElement() {
      this.props.onOpen();
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          title = _props.title,
          actions = _props.actions,
          isOpen = _props.isOpen,
          showHeader = _props.showHeader,
          kind = _props.kind,
          onClose = _props.onClose,
          className = _props.className;

      var classNames = '\n      ' + _defaultStyle2.default[kind] + '\n      ' + (isOpen ? _defaultStyle2.default.isOpen : _defaultStyle2.default.isClosed) + '\n      ' + (showHeader ? _defaultStyle2.default.hasHeader : '') + '\n      ' + (actions && actions.length > 0 ? _defaultStyle2.default.hasFunctions : '') + '\n      ' + className + '\n    ';

      return _react2.default.createElement(
        _StackedEscapable2.default,
        { onEscape: onClose },
        _react2.default.createElement(
          _reactPortal2.default,
          { isOpened: isOpen },
          _react2.default.createElement(
            'div',
            { className: classNames, ref: this.setDialogElement, onClick: this.handleCloseClick },
            _react2.default.createElement(
              'div',
              { className: _defaultStyle2.default.dialog, onClick: this.handleDialogClick },
              !showHeader && onClose && _react2.default.createElement(
                'button',
                { className: _defaultStyle2.default.closeButtonOutside, onClick: onClose },
                _react2.default.createElement(_closeIcon2.default, { color: 'inherit' })
              ),
              _react2.default.createElement(
                'div',
                { className: _defaultStyle2.default.inner },
                showHeader && onClose && _react2.default.createElement(
                  'div',
                  { className: _defaultStyle2.default.header },
                  _react2.default.createElement(
                    'h1',
                    { className: _defaultStyle2.default.title },
                    title
                  ),
                  _react2.default.createElement(
                    'button',
                    { className: _defaultStyle2.default.closeButton, onClick: onClose },
                    _react2.default.createElement(_closeIcon2.default, { color: 'inherit' })
                  )
                ),
                _react2.default.createElement(
                  'div',
                  { className: _defaultStyle2.default.content },
                  this.props.children
                ),
                _react2.default.createElement(
                  'div',
                  { className: _defaultStyle2.default.footer },
                  actions.length > 0 && _react2.default.createElement(
                    'div',
                    { className: _defaultStyle2.default.functions },
                    actions.map(function (action, i) {
                      return _react2.default.createElement(
                        _default2.default,
                        {
                          key: i,
                          onClick: _this2.handleActionClick,
                          'data-action-index': i,
                          color: action.color,
                          disabled: action.disabled,
                          kind: action.kind,
                          autoFocus: action.autoFocus,
                          className: '\n                              ' + _defaultStyle2.default.button + '\n                              ' + (_defaultStyle2.default['button_' + action.kind] || _defaultStyle2.default.button) + '\n                            '
                        },
                        action.title
                      );
                    })
                  )
                )
              )
            )
          )
        )
      );
    }
  }]);

  return DefaultDialog;
}(_react2.default.PureComponent);

DefaultDialog.propTypes = {
  kind: _propTypes2.default.oneOf(['default', 'warning', 'success', 'danger', 'info']),
  className: _propTypes2.default.string,
  title: _propTypes2.default.string.isRequired,
  children: _propTypes2.default.node,
  onOpen: _propTypes2.default.func,
  onClose: _propTypes2.default.func,
  isOpen: _propTypes2.default.bool,
  onAction: _propTypes2.default.func,
  showHeader: _propTypes2.default.bool,
  actions: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    title: _propTypes2.default.string.isRequired,
    tooltip: _propTypes2.default.string,
    kind: _propTypes2.default.string,
    autoFocus: _propTypes2.default.bool
  }))
};
DefaultDialog.defaultProps = {
  isOpen: false,
  showHeader: false,
  onAction: function onAction() {},
  onOpen: function onOpen() {},

  actions: [],
  kind: 'default'
};
exports.default = DefaultDialog;
