// @sanity/form-builder/lib/sanity/inputResolver/customInputs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _boolean = require('part:@sanity/form-builder/input/boolean?');

var _boolean2 = _interopRequireDefault(_boolean);

var _code = require('part:@sanity/form-builder/input/code?');

var _code2 = _interopRequireDefault(_code);

var _datetime = require('part:@sanity/form-builder/input/datetime?');

var _datetime2 = _interopRequireDefault(_datetime);

var _email = require('part:@sanity/form-builder/input/email?');

var _email2 = _interopRequireDefault(_email);

var _geopoint = require('part:@sanity/form-builder/input/geopoint?');

var _geopoint2 = _interopRequireDefault(_geopoint);

var _number = require('part:@sanity/form-builder/input/number?');

var _number2 = _interopRequireDefault(_number);

var _object = require('part:@sanity/form-builder/input/object?');

var _object2 = _interopRequireDefault(_object);

var _reference = require('part:@sanity/form-builder/input/reference?');

var _reference2 = _interopRequireDefault(_reference);

var _richDate = require('part:@sanity/form-builder/input/rich-date?');

var _richDate2 = _interopRequireDefault(_richDate);

var _string = require('part:@sanity/form-builder/input/string?');

var _string2 = _interopRequireDefault(_string);

var _text = require('part:@sanity/form-builder/input/text?');

var _text2 = _interopRequireDefault(_text);

var _url = require('part:@sanity/form-builder/input/url?');

var _url2 = _interopRequireDefault(_url);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// These are inputs that may be implemented by provided parts
exports.default = {
  object: _object2.default,
  boolean: _boolean2.default,
  number: _number2.default,
  string: _string2.default,
  text: _text2.default,
  reference: _reference2.default,
  datetime: _datetime2.default,
  richDate: _richDate2.default,
  email: _email2.default,
  geopoint: _geopoint2.default,
  code: _code2.default,
  url: _url2.default
};
