// @sanity/form-builder/lib/sanity/inputResolver/defaultInputs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Boolean = require('../../inputs/Boolean');

var _Boolean2 = _interopRequireDefault(_Boolean);

var _Email = require('../../inputs/Email');

var _Email2 = _interopRequireDefault(_Email);

var _Number = require('../../inputs/Number');

var _Number2 = _interopRequireDefault(_Number);

var _Object = require('../../inputs/Object');

var _Object2 = _interopRequireDefault(_Object);

var _String = require('../../inputs/String');

var _String2 = _interopRequireDefault(_String);

var _DateTime = require('../../inputs/DateTime');

var _DateTime2 = _interopRequireDefault(_DateTime);

var _Text = require('../../inputs/Text');

var _Text2 = _interopRequireDefault(_Text);

var _Url = require('../../inputs/Url');

var _Url2 = _interopRequireDefault(_Url);

var _SanityArray = require('../inputs/SanityArray');

var _SanityArray2 = _interopRequireDefault(_SanityArray);

var _Image = require('../inputs/Image');

var _Image2 = _interopRequireDefault(_Image);

var _File = require('../inputs/File');

var _File2 = _interopRequireDefault(_File);

var _Slug = require('../inputs/Slug');

var _Slug2 = _interopRequireDefault(_Slug);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  object: _Object2.default,
  array: _SanityArray2.default,
  boolean: _Boolean2.default,
  number: _Number2.default,
  text: _Text2.default,
  email: _Email2.default,
  datetime: _DateTime2.default,
  url: _Url2.default,
  image: _Image2.default,
  file: _File2.default,
  string: _String2.default,
  slug: _Slug2.default
};
