// @sanity/form-builder/lib/inputs/BlockEditor-slate/preview/ListItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ListItem = require('../styles/contentStyles/ListItem.css');

var _ListItem2 = _interopRequireDefault(_ListItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var listItems = ['bullet', 'number', 'roman'];

function ListItem(props) {
  var listItem = props.listItem;
  if (!listItems.includes(listItem)) {
    throw new Error('Don\'t know how to handle listItem \'' + listItem + '\'. ' + ('Expected one of \'' + listItems.join("', '") + '\''));
  }
  var className = _ListItem2.default[listItem] + ' ' + _ListItem2.default['level-' + props.level];
  return _react2.default.createElement(
    'div',
    _extends({}, props.attributes, { className: className }),
    props.contentComponent({ children: props.children })
  );
}

ListItem.propTypes = {
  attributes: _propTypes2.default.object,
  listItem: _propTypes2.default.oneOf(listItems),
  level: _propTypes2.default.number,
  children: _propTypes2.default.node,
  contentComponent: _propTypes2.default.func
};

exports.default = ListItem;
