// @sanity/form-builder/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SlugInput = exports.FileInput = exports.ImageInput = exports.ReferenceInput = exports.BlockEditor = exports.FormBuilderContext = exports.FormBuilder = undefined;

var _FormBuilder = require('./FormBuilder');

Object.defineProperty(exports, 'FormBuilder', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FormBuilder).default;
  }
});

var _FormBuilderContext = require('./FormBuilderContext');

Object.defineProperty(exports, 'FormBuilderContext', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_FormBuilderContext).default;
  }
});

var _BlockEditorSlate = require('./inputs/BlockEditor-slate');

Object.defineProperty(exports, 'BlockEditor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_BlockEditorSlate).default;
  }
});

var _Reference = require('./inputs/Reference');

Object.defineProperty(exports, 'ReferenceInput', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_Reference).default;
  }
});
exports.createFormBuilder = createFormBuilder;

var _File = require('./inputs/File');

var FileInput = _interopRequireWildcard(_File);

var _Image = require('./inputs/Image');

var ImageInput = _interopRequireWildcard(_Image);

var _Slug = require('./inputs/Slug');

var SlugInput = _interopRequireWildcard(_Slug);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Input component factories
exports.ImageInput = ImageInput;
exports.FileInput = FileInput;
exports.SlugInput = SlugInput;
function createFormBuilder() {
  throw new Error('The factory function createFormBuilder(...) has been removed. Please use <FormBuilder .../> instead');
}
