// @sanity/schema/lib/legacy/types/array.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArrayType = undefined;

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OVERRIDABLE_FIELDS = ['jsonType', 'type', 'name', 'title', 'description', 'options', 'fieldsets'];

var ARRAY_CORE = {
  name: 'array',
  type: null,
  jsonType: 'array',
  of: []
};

var ArrayType = exports.ArrayType = {
  get: function get() {
    return ARRAY_CORE;
  },
  extend: function extend(subTypeDef, createMemberType) {
    var parsed = Object.assign((0, _pick3.default)(ARRAY_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: ARRAY_CORE
    });
    (0, _utils.lazyGetter)(parsed, 'of', function () {
      return subTypeDef.of.map(function (ofTypeDef) {
        return createMemberType(ofTypeDef);
      });
    });

    return subtype(parsed);

    function subtype(parent) {
      return {
        get: function get() {
          return parent;
        },

        extend: function extend(extensionDef) {
          if (extensionDef.of) {
            throw new Error('Cannot override `of` property of subtypes of "array"');
          }
          var current = Object.assign({}, parent, (0, _pick3.default)(extensionDef, OVERRIDABLE_FIELDS), { type: parent });
          return subtype(current);
        }
      };
    }
  }
};
