// @sanity/schema/lib/legacy/types/geopoint.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GeoPointType = undefined;

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OVERRIDABLE_FIELDS = ['jsonType', 'type', 'name', 'title', 'description', 'options'];

var GEOPOINT_CORE = {
  name: 'geopoint',
  type: null,
  jsonType: 'object'
};

var GeoPointType = exports.GeoPointType = {
  get: function get() {
    return GEOPOINT_CORE;
  },
  extend: function extend(subTypeDef) {
    var parsed = Object.assign((0, _pick3.default)(GEOPOINT_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: GEOPOINT_CORE
    });

    return subtype(parsed);

    function subtype(parent) {
      return {
        get: function get() {
          return parent;
        },

        extend: function extend(extensionDef) {
          var current = Object.assign({}, parent, (0, _pick3.default)(extensionDef, OVERRIDABLE_FIELDS), { type: parent });
          return subtype(current);
        }
      };
    }
  }
};
