// @sanity/schema/lib/legacy/types/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _any = require('./any');

Object.defineProperty(exports, 'any', {
  enumerable: true,
  get: function get() {
    return _any.AnyType;
  }
});

var _string = require('./string');

Object.defineProperty(exports, 'string', {
  enumerable: true,
  get: function get() {
    return _string.StringType;
  }
});

var _text = require('./text');

Object.defineProperty(exports, 'text', {
  enumerable: true,
  get: function get() {
    return _text.TextType;
  }
});

var _datetime = require('./datetime');

Object.defineProperty(exports, 'datetime', {
  enumerable: true,
  get: function get() {
    return _datetime.DateTimeType;
  }
});

var _date = require('./date');

Object.defineProperty(exports, 'date', {
  enumerable: true,
  get: function get() {
    return _date.DateType;
  }
});

var _boolean = require('./boolean');

Object.defineProperty(exports, 'boolean', {
  enumerable: true,
  get: function get() {
    return _boolean.BooleanType;
  }
});

var _url = require('./url');

Object.defineProperty(exports, 'url', {
  enumerable: true,
  get: function get() {
    return _url.UrlType;
  }
});

var _email = require('./email');

Object.defineProperty(exports, 'email', {
  enumerable: true,
  get: function get() {
    return _email.EmailType;
  }
});

var _number = require('./number');

Object.defineProperty(exports, 'number', {
  enumerable: true,
  get: function get() {
    return _number.NumberType;
  }
});

var _reference = require('./reference');

Object.defineProperty(exports, 'reference', {
  enumerable: true,
  get: function get() {
    return _reference.ReferenceType;
  }
});

var _image = require('./image');

Object.defineProperty(exports, 'image', {
  enumerable: true,
  get: function get() {
    return _image.ImageType;
  }
});

var _file = require('./file');

Object.defineProperty(exports, 'file', {
  enumerable: true,
  get: function get() {
    return _file.FileType;
  }
});

var _object = require('./object');

Object.defineProperty(exports, 'object', {
  enumerable: true,
  get: function get() {
    return _object.ObjectType;
  }
});

var _document = require('./document');

Object.defineProperty(exports, 'document', {
  enumerable: true,
  get: function get() {
    return _document.DocumentType;
  }
});

var _array = require('./array');

Object.defineProperty(exports, 'array', {
  enumerable: true,
  get: function get() {
    return _array.ArrayType;
  }
});

var _slug = require('./slug');

Object.defineProperty(exports, 'slug', {
  enumerable: true,
  get: function get() {
    return _slug.SlugType;
  }
});

var _geopoint = require('./geopoint');

Object.defineProperty(exports, 'geopoint', {
  enumerable: true,
  get: function get() {
    return _geopoint.GeoPointType;
  }
});

var _block = require('./blocks/block');

Object.defineProperty(exports, 'block', {
  enumerable: true,
  get: function get() {
    return _block.BlockType;
  }
});

var _span = require('./blocks/span');

Object.defineProperty(exports, 'span', {
  enumerable: true,
  get: function get() {
    return _span.SpanType;
  }
});
