// @sanity/schema/lib/legacy/types/number.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NumberType = undefined;

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _primitivePreview = require('../preview/primitivePreview');

var _primitivePreview2 = _interopRequireDefault(_primitivePreview);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OVERRIDABLE_FIELDS = ['jsonType', 'type', 'name', 'title', 'description', 'options', 'fieldsets'];

var NUMBER_CORE = {
  name: 'number',
  type: null,
  jsonType: 'number'
};

var NumberType = exports.NumberType = {
  get: function get() {
    return NUMBER_CORE;
  },
  extend: function extend(subTypeDef) {
    var parsed = Object.assign((0, _pick3.default)(NUMBER_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: NUMBER_CORE,
      preview: _primitivePreview2.default
    });

    return subtype(parsed);

    function subtype(parent) {
      return {
        get: function get() {
          return parent;
        },

        extend: function extend(extensionDef) {
          var current = Object.assign({}, parent, (0, _pick3.default)(extensionDef, OVERRIDABLE_FIELDS), { type: parent });
          return subtype(current);
        }
      };
    }
  }
};
