// @sanity/schema/lib/legacy/types/reference.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReferenceType = undefined;

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _arrify = require('arrify');

var _arrify2 = _interopRequireDefault(_arrify);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var OVERRIDABLE_FIELDS = ['jsonType', 'type', 'name', 'title', 'description', 'options'];

var REFERENCE_CORE = {
  name: 'reference',
  type: null,
  jsonType: 'object'
};

function humanize(arr, conjunction) {
  var len = arr.length;
  if (len === 1) {
    return arr[0];
  }
  var first = arr.slice(0, len - 1);
  var last = arr[len - 1];
  return first.join(', ') + ' ' + conjunction + ' ' + last;
}

function buildTitle(type) {
  if (type.title) {
    return type.title;
  }
  if (!type.to || type.to.length === 0) {
    return 'Reference';
  }
  return 'Reference to ' + humanize((0, _arrify2.default)(type.to).map(function (toType) {
    return (toType.title || toType.name || toType.type || '').toLowerCase();
  }), 'or');
}

var ReferenceType = exports.ReferenceType = {
  get: function get() {
    return REFERENCE_CORE;
  },
  extend: function extend(subTypeDef, createMemberType) {
    if (!subTypeDef.to) {
      throw new Error('Missing "to" field in reference definition. Check the type ' + subTypeDef.name);
    }
    var parsed = Object.assign((0, _pick3.default)(REFERENCE_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: REFERENCE_CORE,
      title: subTypeDef.title || buildTitle(subTypeDef)
    });

    (0, _utils.lazyGetter)(parsed, 'to', function () {
      return (0, _arrify2.default)(subTypeDef.to).map(function (toType) {
        return createMemberType(toType);
      });
    });

    return subtype(parsed);

    function subtype(parent) {
      return {
        get: function get() {
          return parent;
        },

        extend: function extend(extensionDef) {
          if (extensionDef.of) {
            throw new Error('Cannot override `of` of subtypes of "reference"');
          }
          var current = Object.assign({}, parent, (0, _pick3.default)(extensionDef, OVERRIDABLE_FIELDS), { type: parent });
          return subtype(current);
        }
      };
    }
  }
};

var CustomRef = ReferenceType.extend({
  name: 'customRef',
  title: 'Custom ref',
  to: []
});

var TypeOfCustomStr = CustomRef.extend({
  name: 'typeOfCustomRef',
  title: 'Type Of CustomRef'
});

var TypeOfTypeOfCustomStr = TypeOfCustomStr.extend({
  name: 'typeOfTypeOfCustomRef',
  title: 'Type Of Type Of CustomRef'
});

_assert2.default.equal(TypeOfTypeOfCustomStr.get().type, TypeOfCustomStr.get());
_assert2.default.equal(TypeOfTypeOfCustomStr.get().type, TypeOfCustomStr.get());
_assert2.default.equal(TypeOfTypeOfCustomStr.get().name, 'typeOfTypeOfCustomRef');
_assert2.default.equal(TypeOfTypeOfCustomStr.get().type.name, 'typeOfCustomRef');
_assert2.default.equal(TypeOfTypeOfCustomStr.get().type.type.name, 'customRef');
_assert2.default.equal(TypeOfTypeOfCustomStr.get().type.type.type.name, 'reference');
_assert2.default.equal(TypeOfTypeOfCustomStr.get().type.type.type.jsonType, 'object');
_assert2.default.equal(TypeOfTypeOfCustomStr.get().to, CustomRef.get().to);
