// @sanity/base/lib/preview/RenderPreviewSnapshot.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _get2 = require('lodash/get');

var _get3 = _interopRequireDefault(_get2);

exports.default = RenderPreviewSnapshot;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _previewResolver = require('part:@sanity/base/preview-resolver?');

var _previewResolver2 = _interopRequireDefault(_previewResolver);

var _SanityDefaultPreview = require('./SanityDefaultPreview');

var _SanityDefaultPreview2 = _interopRequireDefault(_SanityDefaultPreview);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Set this to true for debugging preview subscriptions
var DEBUG = false;

function resolvePreview(type) {
  var fromPreview = (0, _get3.default)(type, 'preview.component');
  if (fromPreview) {
    return fromPreview;
  }
  var custom = _previewResolver2.default && (0, _previewResolver2.default)(type);
  return custom || _SanityDefaultPreview2.default;
}

function RenderPreviewSnapshot(props) {
  var snapshot = props.snapshot,
      type = props.type,
      isLive = props.isLive,
      layout = props.layout;

  var PreviewComponent = resolvePreview(type);
  var preview = _react2.default.createElement(PreviewComponent, {
    value: snapshot,
    type: type,
    layout: layout,
    isPlaceholder: !snapshot
  });
  if (DEBUG) {
    return _react2.default.createElement(
      'div',
      null,
      _react2.default.createElement(
        'span',
        { style: { position: 'absolute', right: 24, top: 2 } },
        isLive ? '⚡️' : '💤'
      ),
      preview
    );
  }

  return preview;
}

RenderPreviewSnapshot.propTypes = {
  snapshot: _propTypes2.default.object,
  type: _propTypes2.default.object,
  isLive: _propTypes2.default.bool,
  layout: _propTypes2.default.string
};
