// @sanity/base/lib/preview/SanityDefaultPreview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _UploadProgressBar = require('./UploadProgressBar');

var _UploadProgressBar2 = _interopRequireDefault(_UploadProgressBar);

var _card = require('part:@sanity/components/previews/card');

var _card2 = _interopRequireDefault(_card);

var _default = require('part:@sanity/components/previews/default');

var _default2 = _interopRequireDefault(_default);

var _detail = require('part:@sanity/components/previews/detail');

var _detail2 = _interopRequireDefault(_detail);

var _inline = require('part:@sanity/components/previews/inline');

var _inline2 = _interopRequireDefault(_inline);

var _media = require('part:@sanity/components/previews/media');

var _media2 = _interopRequireDefault(_media);

var _block = require('part:@sanity/components/previews/block');

var _block2 = _interopRequireDefault(_block);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var previewComponentMap = {
  default: _default2.default,
  card: _card2.default,
  media: _media2.default,
  detail: _detail2.default,
  inline: _inline2.default,
  block: _block2.default
};

function extractUploadState(value) {
  if (!value || (typeof value === 'undefined' ? 'undefined' : _typeof(value)) !== 'object') {
    return { _upload: null, value: value };
  }

  var _upload = value._upload,
      rest = _objectWithoutProperties(value, ['_upload']);

  return { _upload: _upload, value: rest };
}

var SanityDefaultPreview = function (_React$PureComponent) {
  _inherits(SanityDefaultPreview, _React$PureComponent);

  function SanityDefaultPreview() {
    _classCallCheck(this, SanityDefaultPreview);

    return _possibleConstructorReturn(this, (SanityDefaultPreview.__proto__ || Object.getPrototypeOf(SanityDefaultPreview)).apply(this, arguments));
  }

  _createClass(SanityDefaultPreview, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          layout = _props.layout,
          rest = _objectWithoutProperties(_props, ['layout']);

      var PreviewComponent = previewComponentMap.hasOwnProperty(layout) ? previewComponentMap[layout] : previewComponentMap.default;

      var _extractUploadState = extractUploadState(this.props.value),
          _upload = _extractUploadState._upload,
          value = _extractUploadState.value;

      var item = _upload ? _extends({}, value, {
        imageUrl: _upload.previewImage,
        title: value.title || _upload.file && _upload.file.name || 'Uploading…'
      }) : value;

      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(PreviewComponent, _extends({ item: item }, rest, { progress: _upload && _upload.progress }))
      );
    }
  }]);

  return SanityDefaultPreview;
}(_react2.default.PureComponent);

SanityDefaultPreview.propTypes = {
  layout: _propTypes2.default.oneOf(Object.keys(previewComponentMap)),
  value: _propTypes2.default.object
};
exports.default = SanityDefaultPreview;
