// @sanity/desk-tool/lib/utils/draftUtils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.isDraft = isDraft;
exports.isDraftId = isDraftId;
exports.isPublishedId = isPublishedId;
exports.getDraftId = getDraftId;
exports.getPublishedId = getPublishedId;
exports.createDraftFrom = createDraftFrom;
exports.newDraftFrom = newDraftFrom;
exports.createPublishedFrom = createPublishedFrom;
var DRAFTS_FOLDER = exports.DRAFTS_FOLDER = 'drafts';
var DRAFTS_PREFIX = DRAFTS_FOLDER + '.';

function isDraft(document) {
  return isDraftId(document._id);
}

function isDraftId(id) {
  return id.startsWith(DRAFTS_PREFIX);
}

function isPublishedId(id) {
  return !isDraftId(id);
}

function getDraftId(id) {
  return isDraftId(id) ? id : DRAFTS_PREFIX + id;
}

function getPublishedId(id) {
  return isDraftId(id) ? id.slice(DRAFTS_PREFIX.length) : id;
}

function createDraftFrom(document) {
  return _extends({
    _id: getDraftId(document._id)
  }, document);
}

function newDraftFrom(document) {
  return _extends({
    _id: DRAFTS_PREFIX
  }, document);
}

function createPublishedFrom(document) {
  return _extends({
    _id: getPublishedId(document._id)
  }, document);
}
