// @sanity/core/lib/commands/dataset/createDatasetCommand.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _datasetNamePrompt = require('../../actions/dataset/datasetNamePrompt');

var _datasetNamePrompt2 = _interopRequireDefault(_datasetNamePrompt);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

exports.default = {
  name: 'create',
  group: 'dataset',
  signature: '[NAME]',
  description: 'Create a new dataset within your project',
  action: (() => {
    var _ref = _asyncToGenerator(function* (args, context) {
      const apiClient = context.apiClient,
            output = context.output,
            prompt = context.prompt;

      var _args$argsWithoutOpti = _slicedToArray(args.argsWithoutOptions, 1);

      const dataset = _args$argsWithoutOpti[0];

      const client = apiClient();
      const datasetName = yield dataset || (0, _datasetNamePrompt2.default)(prompt);

      try {
        yield client.datasets.create(datasetName);
        output.print('Dataset created successfully');
      } catch (err) {
        throw new Error(`Dataset creation failed:\n${err.message}`);
      }
    });

    return function action(_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })()
};
