// @sanity/base/lib/components/Document.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _generateScriptLoader = require('../util/generateScriptLoader');

var _generateScriptLoader2 = _interopRequireDefault(_generateScriptLoader);

var _AppLoadingScreen = require('./AppLoadingScreen');

var _AppLoadingScreen2 = _interopRequireDefault(_AppLoadingScreen);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// todo: investigate this. Doesn't seem like NODE_ENV gets set on sanity.io
var ENV = process.env.NODE_ENV || 'development';

function assetUrl(staticPath, item) {
  var isAbsolute = item.path.match(/^https?:\/\//);
  if (isAbsolute) {
    return item.path;
  }

  var base = staticPath + '/' + item.path;
  if (!item.hash) {
    return base;
  }

  var hasQuery = base.indexOf('?') !== -1;
  var separator = hasQuery ? '&' : '?';
  return '' + base + separator + item.hash;
}

function Document(props) {
  var stylesheets = props.stylesheets.map(function (item) {
    return _react2.default.createElement('link', {
      key: item.path,
      rel: 'stylesheet',
      href: assetUrl(props.staticPath, item)
    });
  });

  var subresources = props.scripts.map(function (item) {
    return _react2.default.createElement('link', {
      key: item.path,
      rel: 'subresource',
      href: assetUrl(props.staticPath, item)
    });
  });

  var scripts = props.scripts.map(function (item) {
    return assetUrl(props.staticPath, item);
  });
  var scriptLoader = (0, _generateScriptLoader2.default)(scripts);

  var favicons = props.favicons.map(function (item, index) {
    return _react2.default.createElement('link', {
      key: item.path + index,
      rel: 'icon',
      href: assetUrl(props.staticPath, item)
    });
  });
  return _react2.default.createElement(
    'html',
    null,
    _react2.default.createElement(
      'head',
      null,
      _react2.default.createElement('meta', { charSet: props.charset }),
      _react2.default.createElement(
        'title',
        null,
        props.title
      ),
      _react2.default.createElement('meta', { name: 'viewport', content: props.viewport }),
      stylesheets,
      subresources,
      favicons
    ),
    _react2.default.createElement(
      'body',
      null,
      _react2.default.createElement(
        'div',
        { id: 'sanity' },
        _react2.default.createElement(_AppLoadingScreen2.default, { text: props.loading })
      ),
      _react2.default.createElement('script', { dangerouslySetInnerHTML: { __html: scriptLoader } })
    )
  );
}

var asset = _propTypes2.default.shape({
  path: _propTypes2.default.string.isRequired,
  hash: _propTypes2.default.string
});

Document.defaultProps = {
  charset: 'utf-8',
  title: 'Sanity',
  viewport: 'width=device-width, initial-scale=1',
  loading: 'Restoring Sanity…',
  staticPath: '/static',
  favicons: [{ path: 'favicon.ico' }],
  stylesheets: [],
  scripts: []
};

Document.propTypes = {
  charset: _propTypes2.default.string,
  title: _propTypes2.default.string,
  viewport: _propTypes2.default.string,
  loading: _propTypes2.default.node,
  staticPath: _propTypes2.default.string,
  favicons: _propTypes2.default.arrayOf(asset),
  stylesheets: _propTypes2.default.arrayOf(asset),
  scripts: _propTypes2.default.arrayOf(asset)
};

exports.default = Document;
