// @sanity/components/lib/radiobutton/RadioButtonDefault.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _uniqueId2 = require('lodash/uniqueId');

var _uniqueId3 = _interopRequireDefault(_uniqueId2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _defaultStyle = require('part:@sanity/components/radiobutton/default-style');

var _defaultStyle2 = _interopRequireDefault(_defaultStyle);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RadioButton = function (_React$Component) {
  _inherits(RadioButton, _React$Component);

  function RadioButton() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, RadioButton);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = RadioButton.__proto__ || Object.getPrototypeOf(RadioButton)).call.apply(_ref, [this].concat(args))), _this), _this.handleMouseUp = function (event) {
      _this.handleBlur();
    }, _this.handleChange = function () {
      _this.props.onChange(_this.props.item);
    }, _this.handleFocus = function () {
      _this.props.onFocus(_this.props.item);
    }, _this.handleBlur = function () {
      window.setTimeout(function () {
        _this.props.onBlur(_this.props.item);
      }, 0.001);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(RadioButton, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this._inputId = (0, _uniqueId3.default)('RadioSelect');
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          disabled = _props.disabled,
          checked = _props.checked,
          label = _props.label,
          name = _props.name,
          hasFocus = _props.hasFocus;


      return _react2.default.createElement(
        'label',
        {
          className: '\n          ' + _defaultStyle2.default.root + '\n          ' + (disabled ? _defaultStyle2.default.isDisabled : '') + '\n          ' + (checked ? _defaultStyle2.default.isChecked : _defaultStyle2.default.unChecked) + '\n          ' + (hasFocus ? _defaultStyle2.default.isFocused : '') + '\n        ',
          onMouseUp: this.handleMouseUp
        },
        _react2.default.createElement('input', {
          className: _defaultStyle2.default.input,
          type: 'radio',
          onChange: this.handleChange,
          checked: checked,
          id: this._inputId,
          name: name,
          onFocus: this.handleFocus,
          onBlur: this.handleBlur,
          disabled: disabled
        }),
        _react2.default.createElement(
          'div',
          { className: _defaultStyle2.default.label, htmlFor: this._inputId },
          label
        ),
        _react2.default.createElement(
          'div',
          { className: _defaultStyle2.default.circleOutline },
          _react2.default.createElement('div', { className: _defaultStyle2.default.tickOutline })
        ),
        _react2.default.createElement('div', { className: _defaultStyle2.default.focusHelper }),
        _react2.default.createElement('div', { className: _defaultStyle2.default.tickHelper })
      );
    }
  }]);

  return RadioButton;
}(_react2.default.Component);

RadioButton.propTypes = {
  label: _propTypes2.default.string.isRequired,
  item: _propTypes2.default.object.isRequired,
  onChange: _propTypes2.default.func,
  onFocus: _propTypes2.default.func,
  onBlur: _propTypes2.default.func,
  checked: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  name: _propTypes2.default.string,
  hasFocus: _propTypes2.default.bool
};
RadioButton.defaultProps = {
  onChange: function onChange() {},
  onFocus: function onFocus() {},
  onBlur: function onBlur() {}
};
exports.default = RadioButton;
