// @sanity/server/lib/configs/webpack.config.dev.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _webpack3 = require('./webpack.config');

var _webpack4 = _interopRequireDefault(_webpack3);

var _applyStaticLoaderFix = require('../util/applyStaticLoaderFix');

var _applyStaticLoaderFix2 = _interopRequireDefault(_applyStaticLoaderFix);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (config) {
  var baseConfig = (0, _webpack4.default)(config);

  return Object.assign({}, baseConfig, {
    devtool: 'cheap-module-source-map',
    output: Object.assign({ pathinfo: true }, baseConfig.output),
    entry: Object.assign({}, baseConfig.entry, {
      vendor: [require.resolve('eventsource-polyfill'), require.resolve('webpack-hot-middleware/client')].concat(baseConfig.entry.vendor)
    }),
    resolve: {
      alias: Object.assign({}, baseConfig.resolve.alias, {
        'webpack-hot-middleware/client': require.resolve('webpack-hot-middleware/client')
      })
    },
    plugins: (baseConfig.plugins || []).concat([new _webpack2.default.HotModuleReplacementPlugin(), new _webpack2.default.NoEmitOnErrorsPlugin()]),
    module: Object.assign({}, baseConfig.module, {
      rules: (0, _applyStaticLoaderFix2.default)(baseConfig, config)
    })
  });
};
