// @sanity/server/lib/configs/webpack.config.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _resolveFrom = require('resolve-from');

var _resolveFrom2 = _interopRequireDefault(_resolveFrom);

var _v = require('@sanity/webpack-integration/v3');

var _v2 = _interopRequireDefault(_v);

var _extractTextWebpackPlugin = require('extract-text-webpack-plugin');

var _extractTextWebpackPlugin2 = _interopRequireDefault(_extractTextWebpackPlugin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var resolve = function resolve(mod) {
  return require.resolve(mod);
};

exports.default = function () {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var env = config.env || 'development';
  var wpIntegrationOptions = { basePath: config.basePath, env: config.env };
  var basePath = config.basePath || process.cwd();
  var skipMinify = config.skipMinify || false;

  var reactPath = _resolveFrom2.default.silent(basePath, 'react');
  var reactDomPath = _resolveFrom2.default.silent(basePath, 'react-dom');
  var missing = [!reactPath && '`react`', !reactDomPath && '`react-dom`'].filter(Boolean);
  if (!reactPath || !reactDomPath) {
    var missingErr = ['Could not find ' + missing.join(', ') + ' dependencies in project directory', 'These need to be declared in `package.json` and be installed for Sanity to work'].join('\n');

    throw new Error(missingErr);
  }

  var babelConfig = tryRead(_path2.default.join(basePath, '.babelrc'));
  var isProd = env === 'production';

  var cssExtractor = new _extractTextWebpackPlugin2.default({
    filename: 'css/main.css',
    allChunks: true,
    ignoreOrder: true,
    disable: !isProd
  });

  var postcssLoader = {
    loader: resolve('postcss-loader'),
    options: {
      config: {
        path: _path2.default.join(__dirname, 'postcss.config.js')
      }
    }
  };

  var cssLoaderLocation = resolve('css-loader');
  var baseCssLoader = cssLoaderLocation + '?modules&localIdentName=[name]_[local]_[hash:base64:5]&importLoaders=1';
  var cssLoader = isProd && !skipMinify ? baseCssLoader + '&minimize' : baseCssLoader + '&sourceMap';

  var commonChunkPlugin = (typeof config.commonChunkPlugin === 'undefined' || config.commonChunkPlugin) && new _webpack2.default.optimize.CommonsChunkPlugin({ name: 'vendor', filename: 'js/vendor.bundle.js' });

  return {
    entry: {
      app: [!isProd && resolve('react-hot-loader/patch'), resolve('normalize.css'), _path2.default.join(__dirname, '..', 'browser', isProd ? 'entry.js' : 'entry-dev.js')].filter(Boolean),
      vendor: ['react', 'react-dom']
    },
    output: {
      path: config.outputPath || _path2.default.join(__dirname, '..', '..', 'dist'),
      filename: 'js/[name].bundle.js',
      publicPath: '/static/'
    },
    resolve: {
      alias: {
        react: _path2.default.dirname(reactPath),
        'react-dom': _path2.default.dirname(reactDomPath),
        moment$: 'moment/moment.js'
      }
    },
    module: {
      rules: [{
        test: /\.jsx?/,
        exclude: /(node_modules|bower_components)/,
        use: {
          loader: resolve('babel-loader'),
          options: babelConfig || {
            presets: [resolve('babel-preset-react'), resolve('babel-preset-es2015')],
            plugins: [resolve('babel-plugin-syntax-class-properties'), resolve('babel-plugin-transform-class-properties'), !isProd && resolve('react-hot-loader/patch')].filter(Boolean),
            cacheDirectory: true
          }
        }
      }, {
        test: /\.css(\?|$)/,
        use: isProd ? _extractTextWebpackPlugin2.default.extract({ use: [cssLoader, postcssLoader] }) : [resolve('style-loader'), cssLoader, postcssLoader]
      }, {
        test: /\.(jpe?g|png|gif|svg|webp|woff|woff2|ttf|eot|otf)$/,
        use: {
          loader: resolve('file-loader'),
          options: { name: 'assets/[name]-[hash].[ext]' }
        }
      }, _v2.default.getPartLoader(wpIntegrationOptions)]
    },
    profile: config.profile || false,
    plugins: [_v2.default.getEnvPlugin(wpIntegrationOptions), new _webpack2.default.ContextReplacementPlugin(/moment[/\\]locale$/, /en|nb/), _v2.default.getPartResolverPlugin(wpIntegrationOptions), cssExtractor, commonChunkPlugin].filter(Boolean)
  };
};

function tryRead(filePath) {
  try {
    // eslint-disable-next-line no-sync
    var content = _fs2.default.readFileSync(filePath);
    return JSON.parse(content);
  } catch (err) {
    return null;
  }
}
