// @sanity/state-router/lib/components/StateLink.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Link = require('./Link');

var _Link2 = _interopRequireDefault(_Link);

var _internalRouterContextTypeCheck = require('./internalRouterContextTypeCheck');

var _internalRouterContextTypeCheck2 = _interopRequireDefault(_internalRouterContextTypeCheck);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*:: import type {RouterProviderContext} from './types'*/


var EMPTY_STATE = {};

var StateLink = function (_React$PureComponent) {
  _inherits(StateLink, _React$PureComponent);

  function StateLink() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, StateLink);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = StateLink.__proto__ || Object.getPrototypeOf(StateLink)).call.apply(_ref, [this].concat(args))), _this), _this.setElement = function (element /*: ?Link*/) {
      if (element) {
        _this._element = element;
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(StateLink, [{
    key: 'resolveUrl',
    value: function resolveUrl() /*: string*/ {
      var _props = this.props,
          toIndex = _props.toIndex,
          state = _props.state;


      if (state && toIndex) {
        throw new Error('Passing both `state` and `toIndex` as props to StateLink is invalid');
      }

      if (!state && !toIndex) {
        // eslint-disable-next-line no-console
        console.error(new Error('No state passed to StateLink. If you want to link to an empty state, its better to use the the `toIndex` property'));
      }

      var nextState = toIndex ? EMPTY_STATE : state || EMPTY_STATE;

      return this.resolvePathFromState(nextState);
    }
  }, {
    key: 'resolvePathFromState',
    value: function resolvePathFromState(state /*: Object*/) {
      if (!this.context.__internalRouter) {
        return 'javascript://state@' + JSON.stringify(state);
      }
      return this.context.__internalRouter.resolvePathFromState(state);
    }
  }, {
    key: 'focus',
    value: function focus() {
      if (this._element) {
        this._element.focus();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          state = _props2.state,
          toIndex = _props2.toIndex,
          rest = _objectWithoutProperties(_props2, ['state', 'toIndex']);

      return _react2.default.createElement(_Link2.default, _extends({}, rest, { href: this.resolveUrl(), ref: this.setElement }));
    }
  }]);

  return StateLink;
}(_react2.default.PureComponent);

StateLink.defaultProps = {
  replace: false,
  toIndex: false
};
StateLink.contextTypes = {
  __internalRouter: _internalRouterContextTypeCheck2.default
};
exports.default = StateLink;
