// @sanity/desk-tool/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _DeskTool = require('./DeskTool');

var _DeskTool2 = _interopRequireDefault(_DeskTool);

var _viewColumnIcon = require('part:@sanity/base/view-column-icon');

var _viewColumnIcon2 = _interopRequireDefault(_viewColumnIcon);

var _uuid = require('@sanity/uuid');

var _uuid2 = _interopRequireDefault(_uuid);

var _router = require('part:@sanity/base/router');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  router: (0, _router.route)('/:selectedType', [(0, _router.route)('/:action', [(0, _router.route)('/:selectedDocumentId')])]),
  canHandleIntent: function canHandleIntent(intentName, params) {
    return intentName === 'edit' && params.id || intentName === 'create' && params.type;
  },
  getIntentState: function getIntentState(intentName, params) {
    return {
      selectedType: params.type || '*',
      action: 'edit',
      selectedDocumentId: params.id || (0, _uuid2.default)()
    };
  },

  title: 'Desk',
  name: 'desk',
  icon: _viewColumnIcon2.default,
  component: _DeskTool2.default
};
