// @sanity/form-builder/lib/inputs/SearchableStringSelect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _find2 = require('lodash/find');

var _find3 = _interopRequireDefault(_find2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _FormBuilderPropTypes = require('../FormBuilderPropTypes');

var _FormBuilderPropTypes2 = _interopRequireDefault(_FormBuilderPropTypes);

var _default = require('part:@sanity/components/formfields/default');

var _default2 = _interopRequireDefault(_default);

var _searchable = require('part:@sanity/components/selects/searchable');

var _searchable2 = _interopRequireDefault(_searchable);

var _PatchEvent = require('../PatchEvent');

var _PatchEvent2 = _interopRequireDefault(_PatchEvent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SearchableStringSelect = function (_React$Component) {
  _inherits(SearchableStringSelect, _React$Component);

  function SearchableStringSelect() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, SearchableStringSelect);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = SearchableStringSelect.__proto__ || Object.getPrototypeOf(SearchableStringSelect)).call.apply(_ref, [this].concat(args))), _this), _this.handleChange = function (item) {
      _this.props.onChange(_PatchEvent2.default.from((0, _PatchEvent.set)(item.value)));
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(SearchableStringSelect, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          value = _props.value,
          type = _props.type,
          level = _props.level;


      var items = type.options.list.map(function (item) {
        return { title: item, value: item };
      });

      var currentItem = (0, _find3.default)(items, function (item) {
        return item.title === value;
      });

      return _react2.default.createElement(
        _default2.default,
        { label: type.title, level: level, description: type.description },
        _react2.default.createElement(_searchable2.default, {
          placeholder: type.placeholder,
          onChange: this.handleChange,
          onKeyPress: this.handleKeyPress,
          value: currentItem,
          items: items
        })
      );
    }
  }]);

  return SearchableStringSelect;
}(_react2.default.Component);

SearchableStringSelect.propTypes = {
  type: _FormBuilderPropTypes2.default.type.isRequired,
  level: _propTypes2.default.number.isRequired,
  value: _propTypes2.default.string,
  onChange: _propTypes2.default.func
};
SearchableStringSelect.defaultProps = {
  value: '',
  onChange: function onChange() {}
};
exports.default = SearchableStringSelect;
