// @sanity/core/lib/util/pluginChecksumManifest.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setChecksum = setChecksum;
exports.setChecksums = setChecksums;
exports.getChecksum = getChecksum;
exports.getChecksums = getChecksums;
exports.getChecksumsPath = getChecksumsPath;
exports.hasSameChecksum = hasSameChecksum;
exports.localConfigExists = localConfigExists;

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _pathExists = require('path-exists');

var _pathExists2 = _interopRequireDefault(_pathExists);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _normalizePluginName = require('./normalizePluginName');

var _normalizePluginName2 = _interopRequireDefault(_normalizePluginName);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const baseChecksums = {
  '#': 'Used by Sanity to keep track of configuration file checksums, do not delete or modify!'
};

function setChecksum(sanityPath, pluginName, checksum) {
  return getChecksums(sanityPath).then(checksums => {
    checksums[pluginName] = checksum;
    return _fsExtra2.default.writeJson(getChecksumsPath(sanityPath), checksums, { spaces: 2 });
  });
}

function setChecksums(sanityPath, checksums) {
  const sums = Object.assign({}, baseChecksums, checksums);
  return _fsExtra2.default.writeJson(getChecksumsPath(sanityPath), sums, { spaces: 2 });
}

function getChecksum(sanityPath, pluginName) {
  return getChecksums(sanityPath).then(sums => sums[pluginName]);
}

function getChecksums(sanityPath) {
  return _fsExtra2.default.readJson(getChecksumsPath(sanityPath)).catch(() => baseChecksums);
}

function getChecksumsPath(sanityPath) {
  return _path2.default.join(sanityPath, 'config', '.checksums');
}

function hasSameChecksum(sanityPath, pluginName, checksum) {
  return getChecksum(sanityPath, pluginName).then(sum => sum === checksum);
}

function localConfigExists(sanityPath, pluginName) {
  const name = (0, _normalizePluginName2.default)(pluginName);
  return (0, _pathExists2.default)(_path2.default.join(sanityPath, 'config', `${name}.json`));
}
