// @sanity/desk-tool/lib/pane/DocumentsPaneMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _default = require('part:@sanity/components/menus/default');

var _default2 = _interopRequireDefault(_default);

var _barsIcon = require('part:@sanity/base/bars-icon');

var _barsIcon2 = _interopRequireDefault(_barsIcon);

var _thListIcon = require('part:@sanity/base/th-list-icon');

var _thListIcon2 = _interopRequireDefault(_thListIcon);

var _thLargeIcon = require('part:@sanity/base/th-large-icon');

var _thLargeIcon2 = _interopRequireDefault(_thLargeIcon);

var _plusCircleIcon = require('part:@sanity/base/plus-circle-icon');

var _plusCircleIcon2 = _interopRequireDefault(_plusCircleIcon);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }
// import IconSettings from 'part:@sanity/base/cog-icon'


var TEST_CARDS_AND_THUMBNAILS = false;

var LIST_VIEW_ITEMS = [{
  title: 'List',
  icon: _barsIcon2.default,
  action: 'setListLayout',
  key: 'default',
  divider: true
}, {
  title: 'Details',
  icon: _thListIcon2.default,
  action: 'setListLayout',
  key: 'detail'
},
// Disabled for now as it needs more work (virtual-list support, etc.)
TEST_CARDS_AND_THUMBNAILS && {
  title: 'Thumbnails (__DEV__ only)',
  icon: _thLargeIcon2.default,
  action: 'setListLayout',
  key: 'media'

},
// Disabled for now as it needs more work (virtual-list support, etc.)
TEST_CARDS_AND_THUMBNAILS && {
  title: 'Cards (__DEV__ only)',
  icon: _thLargeIcon2.default,
  action: 'setListLayout',
  key: 'card'
}, {
  title: 'Create new…',
  icon: _plusCircleIcon2.default,
  action: 'createNew',
  divider: true
}].filter(Boolean);

var NULL_COMPONENT = function NULL_COMPONENT() {
  return null;
};

var DocumentsPaneMenu = function (_React$PureComponent) {
  _inherits(DocumentsPaneMenu, _React$PureComponent);

  function DocumentsPaneMenu() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, DocumentsPaneMenu);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = DocumentsPaneMenu.__proto__ || Object.getPrototypeOf(DocumentsPaneMenu)).call.apply(_ref, [this].concat(args))), _this), _this.handleMenuAction = function (item) {
      if (item.action === 'setListLayout') {
        _this.props.onSetListLayout(item);
      }

      if (item.action === 'setOrdering') {
        _this.props.onSetOrdering(item.ordering);
      }

      if (item.action === 'createNew') {
        _this.props.onGoToCreateNew();
      }
      _this.props.onMenuClose();
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(DocumentsPaneMenu, [{
    key: 'render',
    value: function render() {
      var orderingOptions = this.props.orderingOptions;

      var orderingItems = orderingOptions.map(function (orderingOption) {
        return {
          title: orderingOption.title,
          icon: orderingOption.icon || NULL_COMPONENT,
          ordering: orderingOption,
          action: 'setOrdering',
          key: orderingOption.name
        };
      }).concat(LIST_VIEW_ITEMS);

      return _react2.default.createElement(_default2.default, _extends({
        onAction: this.handleMenuAction,
        items: orderingItems,
        origin: 'top-right'
      }, this.props));
    }
  }]);

  return DocumentsPaneMenu;
}(_react2.default.PureComponent);

DocumentsPaneMenu.propTypes = {
  onSetListLayout: _propTypes2.default.func,
  onSetOrdering: _propTypes2.default.func,
  onGoToCreateNew: _propTypes2.default.func,
  onMenuClose: _propTypes2.default.func,
  orderingOptions: _propTypes2.default.array
};
exports.default = DocumentsPaneMenu;
