// @sanity/desk-tool/lib/pane/TypePaneItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TypePaneItem = require('./styles/TypePaneItem.css');

var _TypePaneItem2 = _interopRequireDefault(_TypePaneItem);

var _router = require('part:@sanity/base/router');

var _reactInk = require('react-ink');

var _reactInk2 = _interopRequireDefault(_reactInk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TypePaneItem = function (_React$Component) {
  _inherits(TypePaneItem, _React$Component);

  function TypePaneItem() {
    _classCallCheck(this, TypePaneItem);

    return _possibleConstructorReturn(this, (TypePaneItem.__proto__ || Object.getPrototypeOf(TypePaneItem)).apply(this, arguments));
  }

  _createClass(TypePaneItem, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          selected = _props.selected,
          type = _props.type;


      return _react2.default.createElement(
        'div',
        {
          className: selected ? _TypePaneItem2.default.selected : _TypePaneItem2.default.item,
          key: document._id
        },
        _react2.default.createElement(
          _router.StateLink,
          {
            state: { selectedType: type.name },
            className: _TypePaneItem2.default.link
          },
          type.title,
          _react2.default.createElement(_reactInk2.default, { duration: 200, opacity: 0.20, radius: 200 })
        )
      );
    }
  }]);

  return TypePaneItem;
}(_react2.default.Component);

TypePaneItem.propTypes = {
  type: _propTypes2.default.shape({
    name: _propTypes2.default.string,
    title: _propTypes2.default.string,
    icon: _propTypes2.default.element
  }),
  selected: _propTypes2.default.bool
};
exports.default = TypePaneItem;
