// @sanity/components/lib/panes/SplitPaneWrapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SplitPaneWrapper = function (_React$Component) {
  _inherits(SplitPaneWrapper, _React$Component);

  function SplitPaneWrapper() {
    _classCallCheck(this, SplitPaneWrapper);

    return _possibleConstructorReturn(this, (SplitPaneWrapper.__proto__ || Object.getPrototypeOf(SplitPaneWrapper)).apply(this, arguments));
  }

  _createClass(SplitPaneWrapper, [{
    key: 'render',
    value: function render() {
      return this.props.children || _react2.default.createElement('div', null);
    }
  }]);

  return SplitPaneWrapper;
}(_react2.default.Component);

SplitPaneWrapper.propTypes = {
  minWidth: _propTypes2.default.number,
  maxWidth: _propTypes2.default.number,
  defaultWidth: _propTypes2.default.number,
  children: _propTypes2.default.node
};
SplitPaneWrapper.defaultProps = {
  minWidth: 0,
  maxWidth: 0,
  defaultWidth: 0
};
exports.default = SplitPaneWrapper;
