// @sanity/components/lib/fileinput/DefaultFileInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _uniqueId2 = require('lodash/uniqueId');

var _uniqueId3 = _interopRequireDefault(_uniqueId2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FileSelect = function (_React$PureComponent) {
  _inherits(FileSelect, _React$PureComponent);

  function FileSelect() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, FileSelect);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = FileSelect.__proto__ || Object.getPrototypeOf(FileSelect)).call.apply(_ref, [this].concat(args))), _this), _this.handleSelect = function (event) {
      _this.props.onSelect(event.target.files);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(FileSelect, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this._inputId = (0, _uniqueId3.default)('FileSelect');
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      if ('value' in this.props) {
        // eslint-disable-next-line
        console.error(new Error('Warning: A `value` prop was passed to FileSelect. This is most likely not intended as this' + ' component is only for letting users select file(s) from their hard drive'));
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          _props$style = _props.style,
          style = _props$style === undefined ? {} : _props$style,
          className = _props.className,
          onSelect = _props.onSelect,
          rest = _objectWithoutProperties(_props, ['children', 'style', 'className', 'onSelect']);

      return _react2.default.createElement(
        'label',
        {
          style: style,
          className: className,
          htmlFor: this._inputId
        },
        _react2.default.createElement('input', _extends({}, rest, {
          type: 'file',
          value: '',
          id: this._inputId,
          onChange: this.handleSelect,
          style: {
            overflow: 'hidden',
            width: '0.1px',
            height: '0.1px',
            opacity: 0,
            position: 'absolute',
            zIndex: -1
          }
        })),
        children
      );
    }
  }]);

  return FileSelect;
}(_react2.default.PureComponent);

FileSelect.propTypes = {
  onSelect: _propTypes2.default.func.isRequired,
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  style: _propTypes2.default.object
};
FileSelect.defaultProps = {
  onSelect: function onSelect() {}
};
exports.default = FileSelect;
