// @sanity/schema/lib/legacy/types/blocks/span.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpanType = undefined;

var _pick2 = require('lodash/pick');

var _pick3 = _interopRequireDefault(_pick2);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _utils = require('../utils');

var _createPreviewGetter = require('../../preview/createPreviewGetter');

var _createPreviewGetter2 = _interopRequireDefault(_createPreviewGetter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var INHERITED_FIELDS = ['type', 'name', 'title', 'jsonType', 'description', 'options', 'fieldsets'];

var SPAN_CORE = {
  name: 'span',
  type: null,
  jsonType: 'object'
};

var DEFAULT_OPTIONS = {};

var SpanType = {
  get: function get() {
    return SPAN_CORE;
  },
  extend: function extend(subTypeDef, extendMember) {
    var options = _extends({}, subTypeDef.options || DEFAULT_OPTIONS);

    var _subTypeDef$fields = subTypeDef.fields,
        fields = _subTypeDef$fields === undefined ? [] : _subTypeDef$fields,
        _subTypeDef$annotatio = subTypeDef.annotations,
        annotations = _subTypeDef$annotatio === undefined ? [] : _subTypeDef$annotatio,
        _subTypeDef$marks = subTypeDef.marks,
        marks = _subTypeDef$marks === undefined ? [] : _subTypeDef$marks;


    var parsed = Object.assign((0, _pick3.default)(SPAN_CORE, INHERITED_FIELDS), subTypeDef, {
      type: SPAN_CORE,
      options: options
    });

    (0, _utils.lazyGetter)(parsed, 'fields', function () {
      return fields.map(function (fieldDef) {
        var name = fieldDef.name,
            type = _objectWithoutProperties(fieldDef, ['name']);

        return {
          name: name,
          type: extendMember(type)
        };
      });
    });

    (0, _utils.lazyGetter)(parsed, 'annotations', function () {
      return annotations.map(extendMember);
    });
    (0, _utils.lazyGetter)(parsed, 'marks', function () {
      return marks.map(extendMember);
    });

    (0, _utils.lazyGetter)(parsed, 'preview', (0, _createPreviewGetter2.default)(subTypeDef));

    return subtype(parsed);

    function subtype(parent) {
      return {
        get: function get() {
          return parent;
        },

        extend: function extend(extensionDef) {
          if (extensionDef.fields) {
            throw new Error('Cannot override `fields` of subtypes of "span"');
          }
          var current = Object.assign({}, parent, (0, _pick3.default)(extensionDef, INHERITED_FIELDS), { type: parent });
          return subtype(current);
        }
      };
    }
  }
};
exports.SpanType = SpanType;
