// @sanity/server/lib/server.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReactDOM = exports.applyStaticLoaderFix = exports.getDocumentElement = exports.getWebpackProdConfig = exports.getWebpackBaseConfig = exports.getWebpackDevConfig = exports.getWebpackCompiler = exports.getProdServer = exports.getDevServer = undefined;

var _server = require('react-dom/server');

var _server2 = _interopRequireDefault(_server);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _devServer = require('./devServer');

var _devServer2 = _interopRequireDefault(_devServer);

var _prodServer = require('./prodServer');

var _prodServer2 = _interopRequireDefault(_prodServer);

var _webpackConfig = require('./configs/webpack.config.prod');

var _webpackConfig2 = _interopRequireDefault(_webpackConfig);

var _webpackConfig3 = require('./configs/webpack.config.dev');

var _webpackConfig4 = _interopRequireDefault(_webpackConfig3);

var _webpack3 = require('./configs/webpack.config');

var _webpack4 = _interopRequireDefault(_webpack3);

var _applyStaticLoaderFix = require('./util/applyStaticLoaderFix');

var _applyStaticLoaderFix2 = _interopRequireDefault(_applyStaticLoaderFix);

var _baseServer = require('./baseServer');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getWebpackCompiler = function getWebpackCompiler(baseConfig) {
  var config = baseConfig.env === 'production' ? (0, _webpackConfig2.default)(baseConfig) : (0, _webpackConfig4.default)(baseConfig);

  return (0, _webpack2.default)(config);
};

exports.getDevServer = _devServer2.default;
exports.getProdServer = _prodServer2.default;
exports.getWebpackCompiler = getWebpackCompiler;
exports.getWebpackDevConfig = _webpackConfig4.default;
exports.getWebpackBaseConfig = _webpack4.default;
exports.getWebpackProdConfig = _webpackConfig2.default;
exports.getDocumentElement = _baseServer.getDocumentElement;
exports.applyStaticLoaderFix = _applyStaticLoaderFix2.default;
exports.ReactDOM = _server2.default;
