// @sanity/components/lib/previews/DefaultPreview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _defaultStyle = require('part:@sanity/components/previews/default-style');

var _defaultStyle2 = _interopRequireDefault(_defaultStyle);

var _MediaRender = require('./common/MediaRender.js');

var _MediaRender2 = _interopRequireDefault(_MediaRender);

var _SvgPlaceholder = require('./common/SvgPlaceholder');

var _SvgPlaceholder2 = _interopRequireDefault(_SvgPlaceholder);

var _Styleable = require('../utilities/Styleable');

var _Styleable2 = _interopRequireDefault(_Styleable);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PLACEHOLDER = _react2.default.createElement(
  'div',
  { className: _defaultStyle2.default.root },
  _react2.default.createElement(_SvgPlaceholder2.default, { styles: _defaultStyle2.default })
);

var DefaultPreview = function (_React$PureComponent) {
  _inherits(DefaultPreview, _React$PureComponent);

  function DefaultPreview() {
    _classCallCheck(this, DefaultPreview);

    return _possibleConstructorReturn(this, (DefaultPreview.__proto__ || Object.getPrototypeOf(DefaultPreview)).apply(this, arguments));
  }

  _createClass(DefaultPreview, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          item = _props.item,
          assetSize = _props.assetSize,
          emptyText = _props.emptyText,
          children = _props.children,
          isPlaceholder = _props.isPlaceholder,
          progress = _props.progress,
          styles = _props.styles;


      if (!item || isPlaceholder) {
        return _react2.default.createElement(
          'div',
          null,
          PLACEHOLDER
        );
      }

      var hasMedia = item.media || item.sanityImage || item.imageUrl || true;

      return _react2.default.createElement(
        'div',
        {
          className: '\n          ' + styles.root + '\n          ' + (item.subtitle ? styles.hasSubtitle : '') + '\n          ' + (hasMedia ? styles.hasMedia : '') + '\n        '
        },
        hasMedia && _react2.default.createElement(
          'div',
          { className: '' + styles.media },
          _react2.default.createElement(_MediaRender2.default, { size: assetSize, item: item })
        ),
        _react2.default.createElement(
          'div',
          { className: styles.heading },
          _react2.default.createElement(
            'h2',
            { className: styles.title },
            item.title || emptyText
          ),
          item.subtitle && _react2.default.createElement(
            'h3',
            { className: styles.subtitle },
            item.subtitle
          )
        ),
        children && _react2.default.createElement(
          'div',
          { className: styles.children },
          children
        ),
        progress && _react2.default.createElement(
          'div',
          { className: styles.progress },
          _react2.default.createElement('div', { className: styles.progressBar, style: { width: progress + '%' } })
        )
      );
    }
  }]);

  return DefaultPreview;
}(_react2.default.PureComponent);

DefaultPreview.propTypes = {
  item: _propTypes2.default.shape({
    title: _propTypes2.default.string,
    subtitle: _propTypes2.default.string,
    description: _propTypes2.default.string,
    media: _propTypes2.default.node,
    imageUrl: _propTypes2.default.string,
    sanityImage: _propTypes2.default.object
  }),
  assetSize: _propTypes2.default.shape({
    width: _propTypes2.default.number,
    height: _propTypes2.default.number,
    fit: _propTypes2.default.oneOf(['clip', 'crop', 'clamp'])
  }),
  emptyText: _propTypes2.default.string,
  isPlaceholder: _propTypes2.default.bool,
  children: _propTypes2.default.node,
  styles: _propTypes2.default.object,
  progress: _propTypes2.default.number
};
DefaultPreview.defaultProps = {
  emptyText: 'Untitled',
  assetSize: { width: 40, height: 40 },
  progress: undefined,
  item: {}
};
exports.default = (0, _Styleable2.default)(DefaultPreview, _defaultStyle2.default);
