// @sanity/components/lib/previews/DetailPreview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _truncate2 = require('lodash/truncate');

var _truncate3 = _interopRequireDefault(_truncate2);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _detailStyle = require('part:@sanity/components/previews/detail-style');

var _detailStyle2 = _interopRequireDefault(_detailStyle);

var _MediaRender = require('./common/MediaRender.js');

var _MediaRender2 = _interopRequireDefault(_MediaRender);

var _SvgPlaceholder = require('./common/SvgPlaceholder');

var _SvgPlaceholder2 = _interopRequireDefault(_SvgPlaceholder);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var index = 0;

var DetailPreview = function (_React$Component) {
  _inherits(DetailPreview, _React$Component);

  function DetailPreview() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, DetailPreview);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = DetailPreview.__proto__ || Object.getPrototypeOf(DetailPreview)).call.apply(_ref, [this].concat(args))), _this), _this.index = index++, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(DetailPreview, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          item = _props.item,
          emptyText = _props.emptyText,
          assetSize = _props.assetSize,
          children = _props.children,
          isPlaceholder = _props.isPlaceholder;


      if (!item || isPlaceholder) {
        return _react2.default.createElement(
          'div',
          { className: '' + _detailStyle2.default.root },
          _react2.default.createElement(_SvgPlaceholder2.default, { styles: _detailStyle2.default })
        );
      }

      return _react2.default.createElement(
        'div',
        { className: '' + _detailStyle2.default.root },
        _react2.default.createElement(
          'div',
          { className: '' + _detailStyle2.default.media },
          _react2.default.createElement(_MediaRender2.default, { size: assetSize, item: item, fallbackText: 'No media' })
        ),
        _react2.default.createElement(
          'div',
          { className: _detailStyle2.default.content },
          _react2.default.createElement(
            'div',
            { className: _detailStyle2.default.heading },
            _react2.default.createElement(
              'h2',
              { className: _detailStyle2.default.title },
              item.title || emptyText
            ),
            item.subtitle && _react2.default.createElement(
              'h3',
              { className: _detailStyle2.default.subtitle },
              item.subtitle
            )
          ),
          item.description && _react2.default.createElement(
            'p',
            { className: _detailStyle2.default.description },
            (0, _truncate3.default)(item.description, {
              length: 70,
              separator: /,? +/
            })
          )
        ),
        children
      );
    }
  }]);

  return DetailPreview;
}(_react2.default.Component);

DetailPreview.propTypes = {
  item: _propTypes2.default.shape({
    title: _propTypes2.default.string,
    subtitle: _propTypes2.default.string,
    description: _propTypes2.default.string,
    media: _propTypes2.default.node,
    imageUrl: _propTypes2.default.string,
    sanityImage: _propTypes2.default.object
  }),
  assetSize: _propTypes2.default.shape({
    width: _propTypes2.default.number,
    height: _propTypes2.default.number,
    fit: _propTypes2.default.oneOf(['clip', 'crop', 'clamp'])
  }),
  emptyText: _propTypes2.default.string,
  children: _propTypes2.default.node,
  isPlaceholder: _propTypes2.default.bool
};
DetailPreview.defaultProps = {
  emptyText: 'Untitled',
  assetSize: { width: 80, height: 80 }
};
exports.default = DetailPreview;
