// @sanity/components/lib/previews/MediaPreview.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _mediaStyle = require('part:@sanity/components/previews/media-style');

var _mediaStyle2 = _interopRequireDefault(_mediaStyle);

var _MediaRender = require('./common/MediaRender');

var _MediaRender2 = _interopRequireDefault(_MediaRender);

var _SvgPlaceholder = require('./common/SvgPlaceholder');

var _SvgPlaceholder2 = _interopRequireDefault(_SvgPlaceholder);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MediaPreview = function (_React$Component) {
  _inherits(MediaPreview, _React$Component);

  function MediaPreview() {
    _classCallCheck(this, MediaPreview);

    return _possibleConstructorReturn(this, (MediaPreview.__proto__ || Object.getPrototypeOf(MediaPreview)).apply(this, arguments));
  }

  _createClass(MediaPreview, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          item = _props.item,
          emptyText = _props.emptyText,
          assetSize = _props.assetSize,
          children = _props.children,
          aspect = _props.aspect,
          isPlaceholder = _props.isPlaceholder;


      if (!item || isPlaceholder) {
        return _react2.default.createElement(
          'div',
          { className: _mediaStyle2.default.root },
          _react2.default.createElement('div', { className: _mediaStyle2.default.padder, style: { paddingTop: 100 / aspect + '%' } }),
          _react2.default.createElement(
            'div',
            { className: _mediaStyle2.default.mediaContainer },
            _react2.default.createElement(_SvgPlaceholder2.default, { styles: _mediaStyle2.default })
          )
        );
      }

      return _react2.default.createElement(
        'div',
        { className: _mediaStyle2.default.root, title: item.title || emptyText },
        _react2.default.createElement('div', { className: _mediaStyle2.default.padder, style: { paddingTop: 100 / aspect + '%' } }),
        _react2.default.createElement(
          'div',
          { className: _mediaStyle2.default.mediaContainer },
          _react2.default.createElement(_MediaRender2.default, { size: assetSize, item: item, aspect: aspect, fallbackText: item.title || 'No media' })
        ),
        _react2.default.createElement(
          'div',
          { className: '' + _mediaStyle2.default.meta },
          _react2.default.createElement(
            'div',
            { className: '' + _mediaStyle2.default.metaInner },
            item.title && _react2.default.createElement(
              'h2',
              { className: _mediaStyle2.default.title },
              item.title
            ),
            item.subtitle && _react2.default.createElement(
              'h3',
              { className: _mediaStyle2.default.subtitle },
              item.subtitle
            ),
            _react2.default.createElement(
              'p',
              { className: _mediaStyle2.default.description },
              item.description
            )
          )
        ),
        children
      );
    }
  }]);

  return MediaPreview;
}(_react2.default.Component);

MediaPreview.propTypes = {
  item: _propTypes2.default.shape({
    title: _propTypes2.default.string,
    subtitle: _propTypes2.default.string,
    description: _propTypes2.default.string,
    media: _propTypes2.default.node,
    imageUrl: _propTypes2.default.string,
    sanityImage: _propTypes2.default.object
  }),
  assetSize: _propTypes2.default.shape({
    width: _propTypes2.default.number,
    height: _propTypes2.default.number,
    fit: _propTypes2.default.oneOf(['clip', 'crop', 'clamp'])
  }),
  aspect: _propTypes2.default.number,
  emptyText: _propTypes2.default.string,
  children: _propTypes2.default.node,
  isPlaceholder: _propTypes2.default.bool
};
MediaPreview.defaultProps = {
  assetSize: { width: 160, height: 160 },
  aspect: 1,
  emptyText: 'Untitled…'
};
exports.default = MediaPreview;
