// @sanity/form-builder/lib/sanity/SanityFormBuilder.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SanityFormBuilder;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _SanityFormBuilderContext = require('./SanityFormBuilderContext');

var _SanityFormBuilderContext2 = _interopRequireDefault(_SanityFormBuilderContext);

var _FormBuilderInput = require('../FormBuilderInput');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*:: type PatchChannel = {
  subscribe: () => () => {},
  receivePatches: (patches: Array<*>) => void
}*/
/*:: type Props = {
  value: ?any,
  schema: any,
  type: Object,
  patchChannel: PatchChannel,
  onChange: () => {}
}*/
function SanityFormBuilder(props /*: Props*/) {
  return _react2.default.createElement(
    _SanityFormBuilderContext2.default,
    {
      value: props.value,
      schema: props.schema,
      patchChannel: props.patchChannel
    },
    _react2.default.createElement(_FormBuilderInput.FormBuilderInput, {
      type: props.type,
      onChange: props.onChange,
      level: 0,
      value: props.value,
      isRoot: true,
      autoFocus: true
    })
  );
}

SanityFormBuilder.createPatchChannel = _SanityFormBuilderContext2.default.createPatchChannel;
