// @sanity/form-builder/lib/sanity/SanityFormBuilderContext.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SanityFormBuilderContext;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _FormBuilderContext = require('../FormBuilderContext');

var _FormBuilderContext2 = _interopRequireDefault(_FormBuilderContext);

var _preview = require('part:@sanity/base/preview');

var _preview2 = _interopRequireDefault(_preview);

var _inputResolver = require('./inputResolver/inputResolver');

var _inputResolver2 = _interopRequireDefault(_inputResolver);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Default wiring for FormBuilderContext when used as a sanity part
/*:: import type {Node} from 'react'*/


var previewResolver = function previewResolver() {
  return _preview2.default;
};

/*:: type Props = {
  value: ?any,
  schema: Object,
  patchChannel: any,
  children: Node
}*/
function SanityFormBuilderContext(props /*: Props*/) {
  return _react2.default.createElement(
    _FormBuilderContext2.default,
    {
      value: props.value,
      schema: props.schema,
      patchChannel: props.patchChannel,
      resolveInputComponent: _inputResolver2.default,
      resolvePreviewComponent: previewResolver
    },
    props.children
  );
}

SanityFormBuilderContext.createPatchChannel = _FormBuilderContext2.default.createPatchChannel;
