// @sanity/default-layout/lib/datastores/urlState.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.state = undefined;
exports.navigate = navigate;

var _getOrderedTools = require('../util/getOrderedTools');

var _getOrderedTools2 = _interopRequireDefault(_getOrderedTools);

var _defaultLayoutRouter = require('../defaultLayoutRouter');

var _defaultLayoutRouter2 = _interopRequireDefault(_defaultLayoutRouter);

var _location = require('part:@sanity/base/location');

var _location2 = _interopRequireDefault(_location);

var _reconfigureClient = require('../util/reconfigureClient');

var _reconfigureClient2 = _interopRequireDefault(_reconfigureClient);

var _spaces = require('../util/spaces');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function resolveUrlStateWithDefaultSpace(state) {
  if (!_spaces.HAS_SPACES || !state || state.space) {
    return state;
  }
  var defaultSpace = _spaces.CONFIGURED_SPACES.find(function (ds) {
    return ds.default;
  }) || _spaces.CONFIGURED_SPACES[0];
  return Object.assign({}, state, { space: defaultSpace.name });
}

function resolveUrlStateWithDefaultTool(state) {
  if (!state || state.tool) {
    return state;
  }
  return Object.assign({}, state, {
    tool: (0, _getOrderedTools2.default)()[0].name
  });
}

function makeBackwardsCompatible(state) {
  if (!state) {
    return state;
  }
  if ((0, _getOrderedTools2.default)().find(function (tool) {
    return tool.name === state.space;
  })) {
    return Object.assign({}, state, { tool: state.space, space: undefined });
  }
  return state;
}

function resolveDefaultState(state) {
  var urlStateWithDefaultTool = resolveUrlStateWithDefaultTool(makeBackwardsCompatible(state));
  return _spaces.HAS_SPACES ? resolveUrlStateWithDefaultSpace(urlStateWithDefaultTool) : urlStateWithDefaultTool;
}

function resolveIntentState(currentState, intentState) {
  var intent = intentState.intent,
      params = intentState.params;


  var tools = (0, _getOrderedTools2.default)();

  var currentTool = currentState.tool ? tools.find(function (tool) {
    return tool.name === currentState.tool;
  }) : null;

  // If current tool can handle intent and if so, give it precedence
  var matchingTool = (currentTool ? [currentTool].concat(_toConsumableArray(tools)) : tools).find(function (tool) {
    return tool && typeof tool.canHandleIntent === 'function' && tool.canHandleIntent(intent, params);
  });

  if (matchingTool) {
    var toolState = matchingTool.getIntentState(intent, params);
    var currentWithState = resolveUrlStateWithDefaultSpace(currentState) || currentState;
    return Object.assign({}, currentWithState, _defineProperty({
      tool: matchingTool.name
    }, matchingTool.name, toolState));
  }
  return {
    isNotFound: true,
    intent: { name: intent, params: params }
  };
}

function maybeHandleIntent(prevEvent, currentEvent) {
  if (currentEvent.state && currentEvent.state.intent) {
    var redirectState = resolveIntentState(prevEvent ? prevEvent.state : {}, currentEvent.state);
    if (redirectState) {
      navigate(_defaultLayoutRouter2.default.encode(redirectState), { replace: true });
      return null;
    }
  }
  return currentEvent;
}

function decodeUrlState(locationEvent) {
  return {
    type: locationEvent.type,
    state: _defaultLayoutRouter2.default.decode(location.pathname),
    isNotFound: _defaultLayoutRouter2.default.isNotFound(location.pathname)
  };
}

function maybeRedirectDefaultState(event) {
  var redirectState = resolveDefaultState(event.state);
  if (redirectState !== event.state) {
    navigate(_defaultLayoutRouter2.default.encode(redirectState), { replace: true });
    return null;
  }
  return event;
}

function navigate(newUrl, options) {
  _location2.default.actions.navigate(newUrl, options);
}

var state = exports.state = _location2.default.state.map(decodeUrlState).scan(maybeHandleIntent).filter(Boolean).map(maybeRedirectDefaultState).filter(Boolean).publishReplay(1).refCount();

if (_spaces.HAS_SPACES) {
  // Uglybugly mutation ahead.
  state.map(function (event) {
    return event.state;
  }).filter(Boolean).do(_reconfigureClient2.default).subscribe();
}
